---
title: "ChIP mapping and peak finding"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```

# ChIP mapping

- maps the paired end reads to the mouse or combined mouse/human genome as prepared in script 01
- STAR, up to 10000 multimappers, reporting 1 multi mapper at random location

# merge the bam files that were sequenced in multiple libraries and copy all bam files into manuscript folder

```{bash, eval=FALSE, engine="sh"}
module load SAMtools/1.10-foss-2019b
cd ../

bamdir="../michi_110823_160829_2257F1-2615F6_Adnp2_Aparna/bam_files"

samtools merge bam/Adnp2_WT_ut_r3_multi.bam $bamdir/Adnp2_cl10_ChIP_2257F1_multi.bam $bamdir/Adnp2_cl10_ChIP_2615F4_multi.bam
samtools merge bam/Adnp2_WT_ut_r4_multi.bam $bamdir/Adnp2_cl16_ChIP_2257F3_multi.bam $bamdir/Adnp2_cl16_ChIP_2615F5_multi.bam
cp $bamdir/Adnp2_cl10_Input_2615F1_multi.bam bam/Input_WT_ut_r3_multi.bam
cp $bamdir/Adnp2_cl16_Input_2615F2_multi.bam bam/Input_WT_ut_r4_multi.bam

bamdir1="../michi_280923_143751_3534F1-3534F20_Adnp2H3K9me3/bam_files"
bamdir2="../aheljosi_230923_161420_3560F1-3560F6_ChIP_ADNP2_K9_moreReads/bam_files"
bamdir3="../michi_070923_104008_3543F1-3543F7_ADNP2/bam_files"

samtools merge -f -@ 10 bam/Adnp2_WT_ut_r1_multi.bam $bamdir1/A2_589_chip_3534F13_main_multi.bam $bamdir2/A2_589_chip_3560F5_main_multi.bam 
samtools merge -f -@ 10 bam/Adnp2_WT_Setdb1KD_r1_multi.bam $bamdir1/A2_589_dTAG_chip_3534F17_main_multi.bam $bamdir2/A2_589_dTAG_chip_3560F6_main_multi.bam 
cp $bamdir1/A2_589_in_3534F11_main_multi.bam bam/Input_WT_ut_r1_multi.bam

samtools merge -f -@ 10 bam/Adnp2_WT_ut_r2_multi.bam $bamdir1/A2_591_chip_3534F14_main_multi.bam $bamdir3/A2_591_chip_3543F2_main_multi.bam 
samtools merge -f -@ 10 bam/Adnp2_WT_Setdb1KD_r2_multi.bam $bamdir1/A2_591_dTAG_chip_3534F18_main_multi.bam $bamdir3/A2_591_dTAG_chip_3543F3_main_multi.bam 
cp $bamdir1/A2_591_in_3534F12_main_multi.bam bam/Input_WT_ut_r2_multi.bam

cp $bamdir1/A2_D1_chip_3534F15_main_multi.bam bam/Adnp2_PxVxL_ut_r1_multi.bam
cp $bamdir1/A2_D1_dTAG_chip_3534F19_main_multi.bam bam/Adnp2_PxVxL_Setdb1KD_r1_multi.bam
cp $bamdir1/A2_D3_chip_3534F16_main_multi.bam bam/Adnp2_PxVxL_ut_r2_multi.bam
cp $bamdir1/A2_D3_dTAG_chip_3534F20_main_multi.bam bam/Adnp2_PxVxL_Setdb1KD_r2_multi.bam

samtools merge -f -@ 10 bam/H3K9me3_WT_ut_r1_multi.bam $bamdir1/K9_589_chip_3534F3_main_multi.bam $bamdir2/K9_589_chip_3560F1_main_multi.bam 
samtools merge -f -@ 10 bam/H3K9me3_WT_Setdb1KD_r1_multi.bam $bamdir1/K9_589_dTAG_chip_3534F7_main_multi.bam $bamdir2/K9_589_dTAG_chip_3560F3_main_multi.bam 
cp $bamdir1/K9_589_in_3534F1_main_multi.bam bam/InputK9_WT_ut_r1_multi.bam

samtools merge -f -@ 10 bam/H3K9me3_WT_ut_r2_multi.bam $bamdir1/K9_591_chip_3534F4_main_multi.bam $bamdir2/K9_591_chip_3560F2_main_multi.bam 
samtools merge -f -@ 10 bam/H3K9me3_WT_Setdb1KD_r2_multi.bam $bamdir1/K9_591_dTAG_chip_3534F8_main_multi.bam $bamdir2/K9_591_dTAG_chip_3560F4_main_multi.bam 
cp $bamdir1/K9_591_in_3534F2_main_multi.bam bam/InputK9_WT_ut_r2_multi.bam

cp $bamdir1/K9_D1_chip_3534F5_main_multi.bam bam/H3K9me3_PxVxL_ut_r1_multi.bam
cp $bamdir3/K9_D1_dTAG_chip_3543F4_main_multi.bam bam/H3K9me3_PxVxL_Setdb1KD_r1_multi.bam
cp $bamdir1/K9_D3_chip_3534F6_main_multi.bam bam/H3K9me3_PxVxL_ut_r2_multi.bam
cp $bamdir1/K9_D3_dTAG_chip_3534F10_main_multi.bam bam/H3K9me3_PxVxL_Setdb1KD_r2_multi.bam

bamdir="../michi_140823_111744_2472F1-2487F22_Adnp_PxVxL/bam_files"
cp $bamdir/*.bam bam/
cd bam
mv ADNP_FlV5_cl1_FLAG_ChIP_2472F1_multi.bam Adnp_WT_r1_multi.bam
mv ADNP-FlV5_cl1_Input_2487F17_multi.bam InputAdnp_WT_r1_multi.bam
mv ADNP_FlV5_cl6F_FLAG_ChIP_2472F2_multi.bam Adnp_WT_r2_multi.bam
mv ADNP-FlV5_cl6F_Input_2487F18_multi.bam InputAdnp_WT_r2_multi.bam
mv ADNPpxvxl_FlV5_cl1F_FLAG_ChIP_2472F5_multi.bam Adnp_PxVxL_r1_multi.bam
mv ADNPpxvxl-FlV5_cl1F_Input_2487F21_multi.bam InputAdnp_PxVxL_r1_multi.bam
mv ADNPpxvxl_FlV5_cl2F_FLAG_ChIP_2472F6_multi.bam Adnp_PxVxL_r2_multi.bam
mv ADNPpxvxl-FlV5_cl2F_Input_2487F22_multi.bam InputAdnp_PxVxL_r2_multi.bam

bamdir="../michi_110823_153321_1921F1-1921F6_Adnp2_ATAC/bam_files"
cp $bamdir/*.bam bam/
cd bam
mv Adnp2KO_r1_1921F1_multi.bam ATAC_Adnp2KO_r1_multi.bam
mv Adnp2KO_r2_1921F2_multi.bam ATAC_Adnp2KO_r2_multi.bam
mv Adnp2KO_r3_1921F3_multi.bam ATAC_Adnp2KO_r3_multi.bam
mv WT_r1_1921F4_multi.bam ATAC_WT_r1_multi.bam
mv WT_r2_1921F5_multi.bam ATAC_WT_r2_multi.bam
mv WT_r3_1921F6_multi.bam ATAC_WT_r3_multi.bam

#Chd4 and HP1
bamdir="../Luo_Chd4_mES_GencodeGenome/bam_files"
cp $bamdir/ES_Chd4_r1_multi.bam bam
cp $bamdir/ES_Chd4_r2_multi.bam bam

bamdir="../michi_230823_131935_818F8-818F11_HP1b/bam_files"
cp $bamdir/HP1b_wt_ChIP_r1_818F7_multi.bam bam/HP1b_wt_ChIP_r1_multi.bam
cp $bamdir/HP1b_wt_ChIP_r2_818F9_multi.bam bam/HP1b_wt_ChIP_r2_multi.bam
cp $bamdir/HP1b_wt_ChIP_r3_818F11_multi.bam bam/HP1b_wt_ChIP_r3_multi.bam

#index all the bam files
cd bam

for sample in Input_WT_ut_r1 Input_WT_ut_r2 Input_WT_ut_r3 Input_WT_ut_r4  Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4 Adnp2_PxVxL_Setdb1KD_r1 Adnp_PxVxL_r2 ATAC_WT_r1 H3K9me3_WT_Setdb1KD_r1 InputAdnp_WT_r1 Adnp2_PxVxL_Setdb1KD_r2 Adnp_WT_r1 ATAC_WT_r2 H3K9me3_WT_Setdb1KD_r2 InputAdnp_WT_r2 Adnp2_PxVxL_ut_r1 Adnp_WT_r2 ATAC_WT_r3 H3K9me3_WT_ut_r1 InputK9_WT_ut_r1 Adnp2_PxVxL_ut_r2 ATAC_Adnp2KO_r1 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 InputK9_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 ATAC_Adnp2KO_r2 H3K9me3_PxVxL_ut_r1 InputAdnp_PxVxL_r1 Adnp2_WT_Setdb1KD_r2 Adnp_PxVxL_r1 ATAC_Adnp2KO_r3 H3K9me3_PxVxL_ut_r2 InputAdnp_PxVxL_r2 ES_Chd4_r1 ES_Chd4_r2 HP1b_wt_ChIP_r1 HP1b_wt_ChIP_r2 HP1b_wt_ChIP_r3
do
samtools index -@ 10 ${sample}_multi.bam
done

```

# merge the HUMAN spike-in bam files that were sequenced in multiple libraries and copy all bam files into manuscript folder

```{bash, eval=FALSE, engine="sh"}
module load SAMtools/1.10-foss-2019b
cd ../bam/spike

bamdir1="../michi_280923_143751_3534F1-3534F20_Adnp2H3K9me3/bam_files"
bamdir2="../aheljosi_230923_161420_3560F1-3560F6_ChIP_ADNP2_K9_moreReads/bam_files"
bamdir3="../michi_070923_104008_3543F1-3543F7_ADNP2/bam_files"
spikeBed="../mm10_hg38Spike_refSTAR/human_chromosomes.bed"


samtools merge -f -@ 10 Adnp2_WT_ut_r1_multi.bam $bamdir1/A2_589_chip_3534F13_multi.bam $bamdir2/A2_589_chip_3560F5_multi.bam 
samtools merge -f -@ 10 Adnp2_WT_Setdb1KD_r1_multi.bam $bamdir1/A2_589_dTAG_chip_3534F17_multi.bam $bamdir2/A2_589_dTAG_chip_3560F6_multi.bam 
cp $bamdir1/A2_589_in_3534F11_multi.bam Input_WT_ut_r1_multi.bam

samtools merge -f -@ 10 Adnp2_WT_ut_r2_multi.bam $bamdir1/A2_591_chip_3534F14_multi.bam $bamdir3/A2_591_chip_3543F2_multi.bam 
samtools merge -f -@ 10 Adnp2_WT_Setdb1KD_r2_multi.bam $bamdir1/A2_591_dTAG_chip_3534F18_multi.bam $bamdir3/A2_591_dTAG_chip_3543F3_multi.bam 
cp $bamdir1/A2_591_in_3534F12_multi.bam Input_WT_ut_r2_multi.bam

cp $bamdir1/A2_D1_chip_3534F15_multi.bam Adnp2_PxVxL_ut_r1_multi.bam
cp $bamdir1/A2_D1_dTAG_chip_3534F19_multi.bam Adnp2_PxVxL_Setdb1KD_r1_multi.bam
cp $bamdir1/A2_D3_chip_3534F16_multi.bam Adnp2_PxVxL_ut_r2_multi.bam
cp $bamdir1/A2_D3_dTAG_chip_3534F20_multi.bam Adnp2_PxVxL_Setdb1KD_r2_multi.bam

samtools merge -f -@ 10 H3K9me3_WT_ut_r1_multi.bam $bamdir1/K9_589_chip_3534F3_multi.bam $bamdir2/K9_589_chip_3560F1_multi.bam 
samtools merge -f -@ 10 H3K9me3_WT_Setdb1KD_r1_multi.bam $bamdir1/K9_589_dTAG_chip_3534F7_multi.bam $bamdir2/K9_589_dTAG_chip_3560F3_multi.bam 
cp $bamdir1/K9_589_in_3534F1_multi.bam InputK9_WT_ut_r1_multi.bam

samtools merge -f -@ 10 H3K9me3_WT_ut_r2_multi.bam $bamdir1/K9_591_chip_3534F4_multi.bam $bamdir2/K9_591_chip_3560F2_multi.bam 
samtools merge -f -@ 10 H3K9me3_WT_Setdb1KD_r2_multi.bam $bamdir1/K9_591_dTAG_chip_3534F8_multi.bam $bamdir2/K9_591_dTAG_chip_3560F4_multi.bam 
cp $bamdir1/K9_591_in_3534F2_multi.bam InputK9_WT_ut_r2_multi.bam

cp $bamdir1/K9_D1_chip_3534F5_multi.bam H3K9me3_PxVxL_ut_r1_multi.bam
cp $bamdir3/K9_D1_dTAG_chip_3543F4_multi.bam H3K9me3_PxVxL_Setdb1KD_r1_multi.bam
cp $bamdir1/K9_D3_chip_3534F6_multi.bam H3K9me3_PxVxL_ut_r2_multi.bam
cp $bamdir1/K9_D3_dTAG_chip_3534F10_multi.bam H3K9me3_PxVxL_Setdb1KD_r2_multi.bam


#keep only the human reads and index the bam files
for sample in Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_PxVxL_Setdb1KD_r1 H3K9me3_WT_Setdb1KD_r1 Adnp2_PxVxL_Setdb1KD_r2 H3K9me3_WT_Setdb1KD_r2 Adnp2_PxVxL_ut_r1 H3K9me3_WT_ut_r1 Adnp2_PxVxL_ut_r2 H3K9me3_PxVxL_Setdb1KD_r1 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 H3K9me3_PxVxL_ut_r1 Adnp2_WT_Setdb1KD_r2 H3K9me3_PxVxL_ut_r2 InputK9_WT_ut_r1 InputK9_WT_ut_r2 Input_WT_ut_r1 Input_WT_ut_r2
do
samtools index -@ 10 ${sample}_multi.bam
samtools view -b -@ 10 -L ${spikeBed} ${sample}_multi.bam > ${sample}_multi_spike.bam
samtools index -@ 10 ${sample}_multi_spike.bam
rm ${sample}_multi.bam
rm ${sample}_multi.bam.bai
done

```

# count the number of uniquely mapped reads in all samples 

```{bash, eval=FALSE, engine="sh"}
cd ../bam

for sample in Input_WT_ut_r1 Input_WT_ut_r2 Input_WT_ut_r3 Input_WT_ut_r4 Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4  Adnp2_PxVxL_Setdb1KD_r1 Adnp_PxVxL_r2 ATAC_WT_r1 H3K9me3_WT_Setdb1KD_r1 InputAdnp_WT_r1 Adnp2_PxVxL_Setdb1KD_r2 Adnp_WT_r1 ATAC_WT_r2 H3K9me3_WT_Setdb1KD_r2 InputAdnp_WT_r2 Adnp2_PxVxL_ut_r1 Adnp_WT_r2 ATAC_WT_r3 H3K9me3_WT_ut_r1 InputK9_WT_ut_r1 Adnp2_PxVxL_ut_r2 ATAC_Adnp2KO_r1 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 InputK9_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 ATAC_Adnp2KO_r2 H3K9me3_PxVxL_ut_r1 InputAdnp_PxVxL_r1 Adnp2_WT_Setdb1KD_r2 Adnp_PxVxL_r1 ATAC_Adnp2KO_r3 H3K9me3_PxVxL_ut_r2 InputAdnp_PxVxL_r2 ES_Chd4_r1 ES_Chd4_r2 HP1b_wt_ChIP_r1 HP1b_wt_ChIP_r2 HP1b_wt_ChIP_r3
do
nreads=$(samtools view -c -@ 10 -F 4 -q 255 ${sample}_multi.bam) 
printf "%s\t%s\n" ${sample} ${nreads} >> unique_read_counts.txt
done
```

# remove multimappers (-q 255) and non-mappers (-F 4) from ChIPs used for MACS2 peakfinding and index them

```{bash, eval=FALSE, engine="sh"}
for sample in Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4 
do
samtools view -@ 10 -b bam/${sample}_multi.bam -F 4 -q 255 > bam/${sample}_uni.bam
samtools index -@ 10 bam/${sample}_uni.bam
done
```



# peakfinding using MACS2 (remove replicate 5)

```{bash, eval=FALSE, engine="sh"}
cd ..
module purge
module load MACS2/2.2.7.1-foss-2020b-Python-3.8.6
macs2 callpeak -t bam/Adnp2_WT_ut_r1_uni.bam bam/Adnp2_WT_ut_r2_uni.bam bam/Adnp2_WT_ut_r3_uni.bam bam/Adnp2_WT_ut_r4_uni.sampled.bam -f BAMPE -g mm --outdir ../peaks -n Adnp2_WT_ut
```


# calculate ChIP/Input enrichments in peaks and filter out peaks with low enrichments

```{r libraries}
suppressPackageStartupMessages({
  library(tidyverse)
  library(MiniChip)
  library(GenomicFeatures)
  library(Rsamtools)
  library(viridis)
  library(cowplot)
})

#parameters
topdir <-  "../"
datadir <- "data"
genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38Spike_refSTAR/chrNameLength.txt"

```


### calculate GC bias

```{r GCbias}
#| fig-width: 6
#| fig-height: 8
# get genome
mm10 <- readDNAStringSet(genome_file, format="fasta",
                 nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)
names(mm10) <- matrix(unlist(strsplit(names(mm10)," ")),ncol=2,byrow=TRUE)[,1]

#get bam files
bamFiles <- list.files(file.path(topdir,"bam"),pattern = "multi.bam$",full.names = TRUE)
bamFiles <- c(grep("Adnp2_WT_ut_",bamFiles,value=TRUE),grep("Input_WT_ut_",bamFiles,value=TRUE))
bamNames <- gsub("../bam/","",bamFiles)
bamNames <- gsub("_multi.bam","",bamNames)

GCbias_table <- GCbias(
  bamFiles=bamFiles,
  bamNames = bamNames,
  minMQS = 255,
  maxFrag = 1000,
  pe = "both",
  restrict = "chr11",
  winWidth = 500,
  col = inferno,
  genome=mm10,
  GCprob = TRUE,
  span = 0.1,
  plot = FALSE
)


#remove tiles with GC < 0.2 and > 0.7, (and with chip cpm > 4)
GCbias_table2 <- GCbias_table[GCbias_table$bins_gc > 0.2 & GCbias_table$bins_gc < 0.7,]

#model GC bias  
gc <- 100 * GCbias_table2$bins_gc
cpm <- as.matrix(GCbias_table2[,1:10])

gcfit <- lm(cpm ~ gc)
gcslopes <- coefficients(gcfit)["gc", ]

mxx <- range(gc)
mxy <- range(cpm)
pL <- lapply(seq_along(colnames(cpm)), function(ii) {
    i <- colnames(cpm)[ii]
    df <- data.frame(gc = gc, cpm = cpm[, i])
  
    ggplot(df, aes(gc, cpm)) +
        geom_density_2d_filled(bins = 8) +
        scale_fill_manual(values = hcl.colors(8)) +
        geom_smooth(data = df, method = "lm", formula = y~x,
                    linetype = 3, color = "white") +
        coord_cartesian(xlim = mxx, ylim = mxy, expand = FALSE) +
        theme_cowplot(12) +
        labs(x = "G + C (%)",
             y = "log2 CPM + 1") +
        theme(legend.position = "none",
              panel.background = element_rect(fill = "#4B0055")) +
      ggtitle(i)
})
plot_grid(plotlist = pL, nrow = 2)
```


### load peaks and center them arround summit

```{r peaks}
peaks.df <- read.table(file.path(topdir,"peaks/Adnp2_WT_ut_peaks.narrowPeak"),header=F)
names(peaks.df) <- c("chr","start","end","name","score","empty","foldchange","pvalue","qvalue","summit")
peaks <- makeGRangesFromDataFrame(peaks.df,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("chr"),
                                   start.field=c("start"),
                                   end.field=c("end"),
                                   starts.in.df.are.0based=TRUE)
plot(density(width(peaks)))

#remove peaks > 5kb)
peaks <- peaks[width(peaks) < 5000]
plot(density(width(peaks)))

#center peaks on summit
start(peaks) <- start(peaks) + peaks$summit
end(peaks) <- start(peaks)
peaks <- resize(peaks,width=300L, fix="center")
names(peaks) <- peaks$name

#remove peaks that extend over the edge of chromosome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
peaks <- peaks[overlapsAny(peaks,chr_sizes,minoverlap = 300)==TRUE]
common_levels <- seqlevels(peaks)[seqlevels(peaks) %in% seqlevels(chr_sizes)]
peaks <- keepSeqlevels(peaks,value = common_levels)
```

### calculate percent GC in peaks

```{r GC}
#extract peak sequecnes
peak_seqs <- BSgenome::getSeq(mm10,peaks)
#extract GC content
percGC <- as.numeric(Biostrings::letterFrequency(peak_seqs,"GC", as.prob = TRUE)*100)

percGCavg <- mean(percGC)

ggplot(data.frame(gc = percGC), aes(gc)) +
    geom_histogram(bins = nclass.Sturges(percGC),
                   color = "gray30", fill = "gray80") +
    labs(x = "Region G+C (%)",
         y = "Number of regions") +
    geom_vline(xintercept = percGCavg, linetype = 2, col = "firebrick") +
    annotate("text", x = percGCavg, y = Inf, hjust = -0.1, vjust = 1.1,
             label = sprintf("mean = %.1f%%", percGCavg), color = "firebrick") +
    theme_cowplot()
```

### calculate read counts in peaks

```{r counts}
 write.table(data.frame(FileName=bamFiles,SampleName=bamNames),file="QUASR.txt",sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)
    library(QuasR)
    #translate options
    cl <- makeCluster(20)
    Qpaired <- "fr"
    selectReadPosition <- "start"
    orientation <- "any"
    readShiftSize <- "halfInsert"
    minMQS <- 0
    
    #generate project
    proj <- qAlign("QUASR.txt", genome_file, paired = Qpaired, clObj = cl)
    #generate counts matrices
    counts <- qCount(proj, peaks,
                    selectReadPosition= selectReadPosition, orientation = orientation, shift = readShiftSize, 
                    useRead="any", clObj = cl, mapqMin = minMQS)
```

# GC-correct

```{r gc_correct}
cntCorrected <- counts
for (id in setdiff(colnames(counts), "width")) {
    cntCorrected[, id] <- counts[, id] * 2^(gcslopes[id] * (percGCavg - percGC))
}
```

# Visualize GC-dependency before and after correction

```{r vis_before_after_correction}
#| fig-width: 10
#| fig-height: 8

ids <- setdiff(colnames(counts), "width")
mxx <- range(percGC)


    # before correction (top row)
    pL <- lapply(ids[seq(1, length(ids))], function(id) {
        df <- data.frame(gc = percGC,
                         lcnt = log2(counts[, id] + 1))
        mxy <- range(c(log2(counts[, id] + 1), log2(cntCorrected[, id] + 1)))
        ggplot(df, aes(gc, lcnt)) +
            geom_density_2d_filled(bins = 8) +
            scale_fill_manual(values = hcl.colors(8)) +
            geom_smooth(data = df, method = "lm", formula = y~x,
                        linetype = 3, color = "white") +
            coord_cartesian(xlim = mxx, ylim = mxy, expand = FALSE) +
            theme_cowplot(10) +
            labs(x = "G + C (%)",
                 y = "log2 raw count + 1") +
            annotate("text", x = -Inf, y = Inf, hjust = -0.05, vjust = 1.05,
                     label = paste0(id, "\n(raw)"), color = "white") +
            theme(legend.position = "none",
                  panel.background = element_rect(fill = "#4B0055"))
    })
    p1 <- plot_grid(plotlist = pL, ncol = 5, nrow = 2, align = "hv")

    # after correction (bottom row)
    pL <- lapply(ids[seq(1, length(ids))], function(id) {
        df <- data.frame(gc = percGC,
                         lcnt = log2(cntCorrected[, id] + 1))
        mxy <- range(c(log2(counts[, id] + 1), log2(cntCorrected[, id] + 1)))
        ggplot(df, aes(gc, lcnt)) +
            geom_density_2d_filled(bins = 8) +
            scale_fill_manual(values = hcl.colors(8)) +
            geom_smooth(data = df, method = "lm", formula = y~x,
                        linetype = 3, color = "white") +
            coord_cartesian(xlim = mxx, ylim = mxy, expand = FALSE) +
            theme_cowplot(10) +
            labs(x = "G + C (%)",
                 y = "log2 corr. count + 1") +
            annotate("text", x = -Inf, y = Inf, hjust = -0.05, vjust = 1.05,
                     label = paste0(id, "\n(corrected)"), color = "white") +
            theme(legend.position = "none",
                  panel.background = element_rect(fill = "#4B0055"))
    })
    p2 <- plot_grid(plotlist = pL, ncol = 5, nrow = 2, align = "hv")
    
plot_grid(p1, p2, nrow = 2, align = "hv")
```



# select enriched peaks

### normalize counts to total mapped read numbers (cpm)

```{r cpmNormalisation}
normfactor <- alignmentStats(proj)[,"mapped"]/1000000
names(normfactor) <- bamNames
cpmCorrected <- cntCorrected
for (i in seq_along(bamFiles)){
cpmCorrected[,bamNames[i]] <- cntCorrected[,bamNames[i]]/normfactor[bamNames[i]]
}
```

### calculate ChIP/Input enrichment

```{r enrichments}
#| fig-width: 10
#| fig-height: 2

Adnp2_enr <- log2((cpmCorrected[,2:6]+1)/(cpmCorrected[,7:11]+1))
Adnp2_cpmsum <- (cpmCorrected[,2:6])+(cpmCorrected[,7:11])
```

### filter peaks based on enrichment (more than 1.2 fold enriched in at least 3 out of 4 replicates)

```{r filter}
Adnp2_enr1.2 <- Adnp2_enr[,1:4] > log2(1.2)
Adnp2_Peaks_filtered <- peaks[rowSums(Adnp2_enr1.2) >=3]
Adnp2_enr_filtered <- Adnp2_enr[rowSums(Adnp2_enr1.2) >=3,]
```

### save the peaks

```{r}
mcols(Adnp2_Peaks_filtered) <- cbind(mcols(Adnp2_Peaks_filtered),Adnp2_enr_filtered)
saveRDS(object = Adnp2_Peaks_filtered, file = file.path(topdir,datadir, "Adnp2_Peaks_filtered_1.2foldEnr.rds"))
#save as bed file
Adnp2_Peaks_filtered_bed <- data.frame(chr=seqnames(Adnp2_Peaks_filtered),start=start(Adnp2_Peaks_filtered)-1,end=end(Adnp2_Peaks_filtered),
           names=Adnp2_Peaks_filtered$name,
                 scores=Adnp2_Peaks_filtered$score,
                 strands=strand(Adnp2_Peaks_filtered))
write.table(Adnp2_Peaks_filtered_bed, file=file.path(topdir,datadir,"Adnp2_Peaks_filtered_1.2foldEnr.bed"), quote=F, sep="\t", row.names=F, col.names=F)

```

# Session info
```{r session_info}
sessionInfo()
```


