---
title: "Adnp and Adnp2 peak heatmap data"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
suppressPackageStartupMessages({
 library(tidyverse)
library(MiniChip)
library(ComplexHeatmap)
library(GenomicFeatures)
})

#parameters
topdir <-  "../"
datadir <- "data"
plotdir <- "plots"
genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38spike/mm10_hg38Spike_refSTAR/chrNameLength.txt"
repeat_file <-  "../mm10_GRCm38.p6_GencodeM23/mm10_GRCm38.p6.genome_repmask.bed"
gene_file <- "../release_M23/gencode.vM23.annotation.txdb.sqlite"

SEreadShift <- 80

options(ucscChromosomeNames=FALSE)
```

### load peaks

```{r}
ADNP2_peaks <- readRDS(file.path(topdir,datadir,"Adnp2_Peaks_filtered_1.2foldEnr.rds"))
ADNP_peaks <- readRDS(file.path(topdir,datadir,"Adnp_Peaks_filtered_1.2foldEnr.rds"))

peaks <- c(ADNP_peaks,ADNP2_peaks)

#remove peaks that extend over the edge of chromosome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
common_levels <- seqlevels(peaks)[seqlevels(peaks) %in% seqlevels(chr_sizes)]
peaks <- keepSeqlevels(peaks,value = common_levels)
```

### prepare annotations

```{r}

#prepare gene annotations
txdb=loadDb(gene_file)
genes <- genes(txdb)
TSSs <- promoters(genes,upstream=300,downstream=0)
names(TSSs) <- matrix(unlist(strsplit(names(TSSs),".",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]


#prepare repeat annotations from repeat masker run
repeats <- read.delim(repeat_file,header=FALSE)
names(repeats) <- c("chr","start","end","repeat_name","swScore","strand","percDiv","percDel","percIns","genome_left","repeat_class","rep_start","rep_end","rep_left","ID")
repeats <- repeats[complete.cases(repeats),]
reps <- makeGRangesFromDataFrame(repeats,
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("chr"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field=c("strand"),
                                 starts.in.df.are.0based=TRUE)

#remove chromosomes not in combined genome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
reps <- reps[overlapsAny(reps,chr_sizes,minoverlap = 1)==TRUE]

#broad repeat classes
SINEs <- reps[grep("SINE",reps$repeat_class)]
LINEs <- reps[grep("LINE",reps$repeat_class)]
LTRs <- reps[grep("LTR",reps$repeat_class)]

#detailed repeat classes that are enriched in peaks
SINE_B2s <- SINEs[grep("^B2",SINEs$repeat_name)]
SINE_B3s <- SINEs[grep("^B3",SINEs$repeat_name)]
LINE_L1s <- reps[reps$repeat_class=="LINE/L1"]
LTR_ERVKs <- reps[reps$repeat_class=="LTR/ERVK"]

```

# annotate peaks

```{r}
#peaks <- resize(peaks,width=15,fix="center")
peaks$SINE <- ifelse(overlapsAny(peaks,SINEs,ignore.strand=TRUE,minoverlap=10)==TRUE, "SINE","")
peaks$LINE <- ifelse(overlapsAny(peaks,LINEs,ignore.strand=TRUE,minoverlap=10)==TRUE, "LINE","")
peaks$LTR <- ifelse(overlapsAny(peaks,LTRs,ignore.strand=TRUE,minoverlap=10)==TRUE, "LTR","")
peaks$TSS <- ifelse(overlapsAny(peaks,TSSs,ignore.strand=TRUE,minoverlap=10)==TRUE, "TSS","")

peaks$anno <- paste(peaks$SINE,peaks$LINE,peaks$LTR,peaks$TSS,sep="_")
table(peaks$anno)
peaks$anno2 <- ifelse(peaks$SINE=="SINE","SINE",
                      ifelse(peaks$LTR=="LTR","LTR",
                             ifelse(peaks$LINE=="LINE","LINE", "other"
                                    )))
table(peaks$anno2)
saveRDS(peaks,file=file.path(topdir,datadir,"Adnp2_and_Adnp_Peaks_filtered_1.2foldEnr_annotated.rds"))

```



# load ChIP data

```{r}
#get bam files
bamFiles <- list.files(file.path(topdir,"bam"),pattern = "multi.bam$",full.names = TRUE)

bamFiles <- bamFiles[c(1:14, 16:38, 40:45)]

bamFiles <- grep("Input",bamFiles,value=TRUE,invert=TRUE)
bamNames <- gsub("manuscript/bam/","",bamFiles)
bamNames <- gsub("_multi.bam","",bamNames)


#split into paired end and single end bamFiles
bamNamesPE <- c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2","Adnp2_WT_ut_r4",  "Adnp2_WT_ut_r5",
                "Adnp2_PxVxL_ut_r1", "Adnp2_PxVxL_ut_r2","Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2","Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r2","H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2","H3K9me3_PxVxL_ut_r1", "H3K9me3_PxVxL_ut_r2","H3K9me3_WT_Setdb1KD_r1", "H3K9me3_WT_Setdb1KD_r2", "ATAC_WT_r1","ATAC_WT_r2","ATAC_WT_r3","ATAC_Adnp2KO_r1", "ATAC_Adnp2KO_r2","ATAC_Adnp2KO_r3")
bamFilesPE <- paste0(topdir,"/bam/",bamNamesPE,"_multi.bam")

bamNamesSE <- c("Adnp_WT_r1","Adnp_WT_r2","Adnp_PxVxL_r1","Adnp_PxVxL_r2","ES_Chd4_r1","ES_Chd4_r2","HP1b_wt_ChIP_r1","HP1b_wt_ChIP_r2","HP1b_wt_ChIP_r3")
bamFilesSE <- paste0(topdir,"/bam/",bamNamesSE,"_multi.bam")
```


# calculate heatmap counts

```{r}

span = 1500
step = 15

countsSE <- SummitHeatmap(
  peaks,
  bamFilesSE,
  bamNamesSE,
  span = span,
  step = step,
  minOverlap = 1,
  useCPM = TRUE,
  PairedEnd = FALSE,
  minMQS = 255,
  strand = 0,
  readExtension3 = 0,
  readShiftSize = SEreadShift,
  read2pos = 5,
  mode="Q",
  genome=genome_file
)

countsPE <- SummitHeatmap(
  peaks,
  bamFilesPE,
  bamNamesPE,
  span = span,
  step = step,
  minOverlap = 1,
  useCPM = TRUE,
  PairedEnd = TRUE,
  minMQS = 255,
  strand = 0,
  readExtension3 = 0,
  readShiftSize = "halfInsert",
  read2pos = 5,
  mode="Q",
  genome=genome_file
)

counts <- c(countsSE,countsPE)

```

#summarize the counts

```{r}
sampleList <- list(Adnp2_WT=c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2", "Adnp2_WT_ut_r4","Adnp2_WT_ut_r5"),
                   Adnp2_WT_ut=c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2"),
                   Adnp2_PxVxL_ut=c("Adnp2_PxVxL_ut_r1", "Adnp2_PxVxL_ut_r2"),
                   Adnp2_WT_Setdb1KD=c("Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2"),
                   Adnp2_PxVxL_Setdb1KD=c("Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r2"),
                   Adnp_WT=c("Adnp_WT_r1","Adnp_WT_r2"),
                   Adnp_PxVxL=c("Adnp_PxVxL_r1","Adnp_PxVxL_r2"),
                   Chd4=c("ES_Chd4_r1","ES_Chd4_r2"),
                   HP1b=c("HP1b_wt_ChIP_r1","HP1b_wt_ChIP_r2","HP1b_wt_ChIP_r3"),
                   H3K9me3_WT_ut=c("H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2"),
                   H3K9me3_PxVxL_ut=c("H3K9me3_PxVxL_ut_r1", "H3K9me3_PxVxL_ut_r2"),
                   H3K9me3_WT_Setdb1KD=c("H3K9me3_WT_Setdb1KD_r1", "H3K9me3_WT_Setdb1KD_r2"),
                   ATAC_WT=c("ATAC_WT_r1","ATAC_WT_r2","ATAC_WT_r3"),
                   ATAC_Adnp2KO=c("ATAC_Adnp2KO_r1", "ATAC_Adnp2KO_r2","ATAC_Adnp2KO_r3")
)

#summarize counts per group
counts2 <- SummarizeHeatmaps(counts, sampleList=sampleList, summarizing = mean)
```

# annotation heatmap calculations

```{r}
TSSs.anno <- AnnotationHeatmap(
  peaks,
  TSSs,
  annoname = "TSSs",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)
SINEs.anno <- AnnotationHeatmap(
  peaks,
  SINEs,
  annoname = "SINEs",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)
LINEs.anno <- AnnotationHeatmap(
  peaks,
  LINEs,
  annoname = "LINEs",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)
LTRs.anno <- AnnotationHeatmap(
  peaks,
  LTRs,
  annoname = "LTRs",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)
SINE_B2s.anno <- AnnotationHeatmap(
  peaks,
  SINE_B2s,
  annoname = "SINE_B2s",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)
SINE_B3s.anno <- AnnotationHeatmap(
  peaks,
  SINE_B3s,
  annoname = "SINE_B3s",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)
LINE_L1s.anno <- AnnotationHeatmap(
  peaks,
  LINE_L1s,
  annoname = "LINE_L1s",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)
LTR_ERVKs.anno <- AnnotationHeatmap(
  peaks,
  LTR_ERVKs,
  annoname = "LTR_ERVKs",
  span = span,
  step = step,
  ignoreStrand = TRUE,
  minoverlap = round(step/2)
)

#combine annotation counts
annos <- list(TSSs.anno,SINEs.anno,LINEs.anno,LTRs.anno,SINE_B2s.anno,SINE_B3s.anno,LINE_L1s.anno,LTR_ERVKs.anno)
names(annos) <- c("TSSs","SINEs","LINEs","LTRs","SINE_B2s","SINE_B3s","LINE_L1s","LTR_ERVKs")

#combine counts and annotation heatmap counts and save
counts3 <- c(counts2,annos)
saveRDS(counts3,file.path(topdir,datadir,sprintf("heatmap_counts_at_Adnp_and_Adnp2_peaks_SEreadShift_%d.rds",SEreadShift)))
```


# calculate heatmap counts with read extension instead of shifting

```{r}

span = 1500
step = 15

countsSE <- SummitHeatmap(
  peaks,
  bamFilesSE,
  bamNamesSE,
  span = span,
  step = step,
  minOverlap = 1,
  useCPM = TRUE,
  PairedEnd = FALSE,
  minMQS = 255,
  strand = 0,
  readExtension3 = 150,
  readShiftSize = 0,
  read2pos = 0,
  mode="F"
)

countsPE <- SummitHeatmap(
  peaks,
  bamFilesPE,
  bamNamesPE,
  span = span,
  step = step,
  minOverlap = 1,
  useCPM = TRUE,
  PairedEnd = TRUE,
  minMQS = 255,
  strand = 0,
  readExtension3 = 150,
  readShiftSize = 0,
  read2pos = 0,
  mode="F"
)

counts <- c(countsSE,countsPE)

```

#summarize the counts

```{r}
#summarize counts per group
counts2 <- SummarizeHeatmaps(counts, sampleList=sampleList, summarizing = mean)
```

#combine with annos and save

```{r}
counts4 <- c(counts2,annos)
saveRDS(counts4,file.path(topdir,datadir,"heatmap_counts_at_Adnp_and_Adnp2_peaks_readExtension150.rds"))
```

