---
title: "Adnp and Adnp2 repeat start heatmap data"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
suppressPackageStartupMessages({
 library(tidyverse)
library(edgeR)
library(ComplexHeatmap)
library(GenomicFeatures)
library(Rsubread)  
  library(cowplot)
  library(MiniChip)
})

#parameters
topdir <-  "../"
datadir <- "data"
plotdir <- "plots"
genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38Spike_refSTAR/chrNameLength.txt"
repeat_file <-  "../mm10_GRCm38.p6_GencodeM23/mm10_GRCm38.p6.genome_repmask.bed"
gene_file <- "../release_M23/gencode.vM23.annotation.txdb.sqlite"

SEreadShift <- 80
minMQS=0

options(ucscChromosomeNames=FALSE)

```



### prepare annotations

```{r}

#prepare repeat annotations from repeat masker run
repeats <- read.delim(repeat_file,header=FALSE)
names(repeats) <- c("chr","start","end","repeat_name","swScore","strand","percDiv","percDel","percIns","genome_left","repeat_class","rep_start","rep_end","rep_left","ID")
repeats <- repeats[complete.cases(repeats),]
reps <- makeGRangesFromDataFrame(repeats,
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("chr"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field=c("strand"),
                                 starts.in.df.are.0based=TRUE)

#remove chromosomes not in combined genome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
reps <- reps[overlapsAny(reps,chr_sizes,minoverlap = 1)==TRUE]
common_levels <- seqlevels(reps)[seqlevels(reps) %in% seqlevels(chr_sizes)]
reps <- keepSeqlevels(reps,value = common_levels)
#add a unique names column
names(reps) <- paste(reps$repeat_name,seqnames(reps),start(reps),end(reps),sep="_")

#select the repeat names that are peak enriched (Adnp or Adnp2)
Adnp2_repoverlap <- read.table(file.path(topdir,datadir,"repeat_names_at_Adnp2_vs_random_peaks.txt"),sep="\t",header=TRUE)
Adnp_repoverlap <- read.table(file.path(topdir,datadir,"repeat_names_at_Adnp_vs_random_peaks.txt"),sep="\t",header=TRUE)

Adnp2_repoverlap$peak_random <- Adnp2_repoverlap$Adnp2/Adnp2_repoverlap$random
Adnp_repoverlap$peak_random <- Adnp_repoverlap$Adnp/Adnp_repoverlap$random

enr_rep_names <- c(Adnp2_repoverlap$repeat_name[Adnp2_repoverlap$peak_random > 5 & Adnp2_repoverlap$Adnp2 > 100],Adnp_repoverlap$repeat_name[Adnp_repoverlap$peak_random > 5 & Adnp_repoverlap$Adnp > 1000])
enr_reps <- reps[reps$repeat_name %in% enr_rep_names]

#add a unique names column
names(enr_reps) <- paste(enr_reps$repeat_name,seqnames(enr_reps),start(enr_reps),end(enr_reps),sep="_")

#save for later use
#saveRDS(enr_reps,file.path(topdir,datadir,"repeats_with_ADNP_ADNP2_paks.rds")) 
```

# add random regions in the genome (same number and length as IAPEz-int) as control

```{r}
random_reps <- SimulatePeaks(length(reps[reps$repeat_name=="IAPEz-int"]),peak.widths = width(reps[reps$repeat_name=="IAPEz-int"]),chromosomeSizes = chrsizes_file)
random_reps$repeat_name <- "random"
random_reps$swScore <- 0
random_reps$repeat_class <- "random"

mcols(reps) <- mcols(reps)[,c(1,2,7)]
reps <- c(reps,random_reps)
```


# load ChIP data

```{r}
#get bam files
bamFiles <- list.files(file.path(topdir,"bam"),pattern = "multi.bam$",full.names = TRUE)

#split into paired end and single end bamFiles
bamNamesPE <- c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2", "Adnp2_WT_ut_r3","Adnp2_WT_ut_r4",
                "Adnp2_PxVxL_ut_r1", "Adnp2_PxVxL_ut_r2","Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2","Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r2","H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2","H3K9me3_PxVxL_ut_r1", "H3K9me3_PxVxL_ut_r2","H3K9me3_WT_Setdb1KD_r1", "H3K9me3_WT_Setdb1KD_r2","H3K9me3_PxVxL_Setdb1KD_r1","H3K9me3_PxVxL_Setdb1KD_r2", "ATAC_WT_r1","ATAC_WT_r2","ATAC_WT_r3","ATAC_Adnp2KO_r1", "ATAC_Adnp2KO_r2","ATAC_Adnp2KO_r3",
                "Input_WT_ut_r1","Input_WT_ut_r2","Input_WT_ut_r3","Input_WT_ut_r4",
                "InputK9_WT_ut_r1","InputK9_WT_ut_r2")
bamFilesPE <- paste0("manuscript/bam/",bamNamesPE,"_multi.bam")

bamNamesSE <- c("Adnp_WT_r1","Adnp_WT_r2","Adnp_PxVxL_r1","Adnp_PxVxL_r2","ES_Chd4_r1","ES_Chd4_r2","HP1b_wt_ChIP_r1","HP1b_wt_ChIP_r2","HP1b_wt_ChIP_r3",
               "InputAdnp_WT_r1","InputAdnp_WT_r2","InputAdnp_PxVxL_r1","InputAdnp_PxVxL_r2" )
bamFilesSE <- paste0("manuscript/bam/",bamNamesSE,"_multi.bam")
```


# calculate read counts in repeats for ChIP and Input

```{r}
saf <- data.frame(GeneID= names(reps), Chr=seqnames(reps),
                  Start=start(reps), End=end(reps),Strand=strand(reps))

f_countsSE <- featureCounts(bamFilesSE,annot.ext=saf,useMetaFeatures=FALSE,allowMultiOverlap=TRUE,
                            minOverlap=1,countMultiMappingReads=TRUE,fraction=FALSE,
                            minMQS=minMQS,strandSpecific=0,nthreads=10,verbose=FALSE,
                            isPairedEnd=FALSE,
                            readExtension3 = 150,read2pos = NULL
)
f_countsPE <- featureCounts(bamFilesPE,annot.ext=saf,useMetaFeatures=FALSE,allowMultiOverlap=TRUE,
                            minOverlap=1,countMultiMappingReads=TRUE,fraction=FALSE,
                            minMQS=minMQS,strandSpecific=0,nthreads=10,verbose=FALSE,
                            isPairedEnd=TRUE,
                            readExtension3 = 150,read2pos = NULL
)

counts <- cbind(f_countsSE$counts,f_countsPE$counts)
colnames(counts) <- c(bamNamesSE,bamNamesPE)
mapped.reads <- c(apply(f_countsSE$stat[c(1,12),-1],2,sum),apply(f_countsPE$stat[c(1,12),-1],2,sum))
names(mapped.reads) <- c(bamNamesSE,bamNamesPE)
```

# calculate cpm 

```{r}
#cpm
d0 <- DGEList(counts,lib.size=mapped.reads)
d0 <- calcNormFactors(d0,method="none")
cpms <- data.frame(cpm(d0))
```

# calculate fpkm

```{r}
#fpkm
replenth <- width(reps)/1000
fpkms <- data.frame(cpm(d0)/replenth)
fpkms$ID <- row.names(fpkms)
```

# add repeat info to cpms

```{r}
#add repeat info to cpms
reps$ID <- names(reps)
reps$repclass <- ifelse(grepl("^DNA",reps$repeat_class),"DNA",
                              ifelse(grepl("RNA",reps$repeat_class),"RNA",
                                     ifelse(grepl("^LINE",reps$repeat_class),"LINE",
                                         ifelse(grepl("^LTR",reps$repeat_class),"LTR",
                                                ifelse(grepl("^SINE",reps$repeat_class),"SINE",
                                                       ifelse(grepl("Satellite",reps$repeat_class),"Satellite",
                                                              ifelse(reps$repeat_class=="Simple_repeat" | reps$repeat_class=="Low_complexity","Simple/low_complexity",
                                                                     ifelse(reps$repeat_class=="random","random","other"))))))))
cpms2 <- cbind(cpms,data.frame(mcols(reps)))
fpkms2 <- cbind(fpkms,data.frame(mcols(reps)))

saveRDS(cpms2,file.path(topdir,datadir,"per_repeat_per_sample_cpms.rds"))
saveRDS(fpkms2,file.path(topdir,datadir,"per_repeat_per_sample_fpkms_TMM.rds"))

```

# summarize per repeat family: sum of cpms per repeat family, mean and sd across replicates, save file

```{r}
cpms2 <- readRDS(file.path(topdir,datadir,"per_repeat_per_sample_cpms.rds"))

cpms2sums <- cpms2 %>% group_by(repclass,repeat_class,repeat_name) %>% summarize_if(is.numeric, sum, na.rm = TRUE)

cpms2sumsAves <- data.frame(repclass=cpms2sums$repclass,repeat_class=cpms2sums$repeat_class,repeat_name=cpms2sums$repeat_name,
                            Adnp_WT=apply(cpms2sums[,c("Adnp_WT_r1","Adnp_WT_r2")],1,mean),
                            Adnp_PxVxL=apply(cpms2sums[,c("Adnp_PxVxL_r1","Adnp_PxVxL_r2")],1,mean),
                            
                            Adnp2_WT_ut=apply(cpms2sums[,c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2","Adnp2_WT_ut_r3","Adnp2_WT_ut_r4")],1,mean),

                            Adnp2_PxVxL_ut=apply(cpms2sums[,c("Adnp2_PxVxL_ut_r1","Adnp2_PxVxL_ut_r2")],1,mean),
                            Adnp2_WT_Setdb1KD=apply(cpms2sums[,c("Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2")],1,mean),
                            Adnp2_PxVxL_Setdb1KD=apply(cpms2sums[,c("Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r2")],1,mean),
                            
                            H3K9me3_WT_ut=apply(cpms2sums[,c("H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2")],1,mean),
                            H3K9me3_WT_Setdb1KD=apply(cpms2sums[,c("H3K9me3_WT_Setdb1KD_r1","H3K9me3_WT_Setdb1KD_r2")],1,mean),
                            H3K9me3_PxVxL_Setdb1KD=apply(cpms2sums[,c("H3K9me3_PxVxL_Setdb1KD_r1","H3K9me3_PxVxL_Setdb1KD_r2")],1,mean),
                            H3K9me3_PxVxL_ut=apply(cpms2sums[,c("H3K9me3_PxVxL_ut_r1","H3K9me3_PxVxL_ut_r2")],1,mean),

                            Adnp_Input=apply(cpms2sums[,c("InputAdnp_WT_r1","InputAdnp_WT_r2")],1,mean),
                            Adnp2_Input=apply(cpms2sums[,c("Input_WT_ut_r1","Input_WT_ut_r2","Input_WT_ut_r3","Input_WT_ut_r4")],1,mean),

                            H3K9me3_Input=apply(cpms2sums[,c("InputK9_WT_ut_r1","InputK9_WT_ut_r2")],1,mean))

cpms2sumsSDs <- data.frame(repclass=cpms2sums$repclass,repeat_class=cpms2sums$repeat_class,repeat_name=cpms2sums$repeat_name,
                             Adnp_WT=apply(cpms2sums[,c("Adnp_WT_r1","Adnp_WT_r2")],1,sd),
                            Adnp_PxVxL=apply(cpms2sums[,c("Adnp_PxVxL_r1","Adnp_PxVxL_r2")],1,sd),
                           
                            Adnp2_WT_ut=apply(cpms2sums[,c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2","Adnp2_WT_ut_r3","Adnp2_WT_ut_r4")],1,sd),

                            Adnp2_PxVxL_ut=apply(cpms2sums[,c("Adnp2_PxVxL_ut_r1","Adnp2_PxVxL_ut_r2")],1,sd),
                            Adnp2_WT_Setdb1KD=apply(cpms2sums[,c("Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2")],1,sd),
                            Adnp2_PxVxL_Setdb1KD=apply(cpms2sums[,c("Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r1")],1,sd),
                            
                            H3K9me3_WT_ut=apply(cpms2sums[,c("H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2")],1,sd),
                            H3K9me3_WT_Setdb1KD=apply(cpms2sums[,c("H3K9me3_WT_Setdb1KD_r1","H3K9me3_WT_Setdb1KD_r2")],1,sd),
                            H3K9me3_PxVxL_Setdb1KD=apply(cpms2sums[,c("H3K9me3_PxVxL_Setdb1KD_r1","H3K9me3_PxVxL_Setdb1KD_r2")],1,sd),
                            H3K9me3_PxVxL_ut=apply(cpms2sums[,c("H3K9me3_PxVxL_ut_r1","H3K9me3_PxVxL_ut_r2")],1,sd),


                            Adnp_Input=apply(cpms2sums[,c("InputAdnp_WT_r1","InputAdnp_WT_r2")],1,sd),
                            Adnp2_Input=apply(cpms2sums[,c("Input_WT_ut_r1","Input_WT_ut_r2","Input_WT_ut_r3","Input_WT_ut_r4")],1,sd),

                            H3K9me3_Input=apply(cpms2sums[,c("InputK9_WT_ut_r1","InputK9_WT_ut_r2")],1,sd)
                            )

saveRDS(cpms2sumsAves,file.path(topdir,datadir,"per_repeat_family_per_group_mean_cpms.rds"))
saveRDS(cpms2sumsSDs,file.path(topdir,datadir,"per_repeat_family_per_group_sd_cpms.rds"))
```


# barplots 

### put mean and sd tables in long format and combine

```{r}
cpms2sumsAves <- readRDS(file.path(topdir,datadir,"per_repeat_family_per_group_mean_cpms.rds"))
cpms2sumsSDs <- readRDS(file.path(topdir,datadir,"per_repeat_family_per_group_sd_cpms.rds"))

cpms2sumsAves_long <- cpms2sumsAves[,1:16] %>% pivot_longer(cols=colnames(cpms2sumsAves[,1:16])[4:ncol(cpms2sumsAves[,1:16])],names_to = "sample",values_to="cpm")
cpms2sumsSDs_long <- cpms2sumsSDs %>% pivot_longer(cols=colnames(cpms2sumsSDs)[4:ncol(cpms2sumsSDs)],names_to = "sample",values_to="cpm_sd")
cpms2sumsAves_long <- cbind(cpms2sumsAves_long,cpms2sumsSDs_long[,5])

```


# barplots of individual enriched interesting repeats, including replicate points. split by ChIP type.

```{r}
cpms2sumsAves_long_sel <- cpms2sumsAves_long[cpms2sumsAves_long$repeat_name %in% c("IAPEz-int","MMERVK10C-int","RLTR10C","IAPLTR1a_Mm","L1MdA_I"),]
cpms2sumsAves_long_sel$ChIP <- ifelse(grepl("^H3K9me3",cpms2sumsAves_long_sel$sample),"H3K9me3",
                                      ifelse(grepl("^Adnp2",cpms2sumsAves_long_sel$sample),"ADNP2","ADNP"))
  
##########individual replicate data for dots##############

cpms3sums_long  <- cpms2sums[cpms2sums$repeat_name %in% c("IAPEz-int","MMERVK10C-int","RLTR10C","IAPLTR1a_Mm","L1MdA_I"),] %>% pivot_longer(cols=colnames(cpms2sums)[4:ncol(cpms2sums)],names_to = "sample",values_to="cpm")

#remove the extra (non-matched) replicates from ADNP2 data
cpms3sums_long <- cpms3sums_long[grepl("_r[0-2]",cpms3sums_long$sample),]
cpms3sums_long$group <- gsub("_r[0-9]","",cpms3sums_long$sample)
cpms3sums_long$group <- ifelse(cpms3sums_long$group=="InputK9_WT_ut","H3K9me3_Input",
                               ifelse(cpms3sums_long$group=="Input_WT_ut","Adnp2_Input",
                                      ifelse(cpms3sums_long$group=="InputAdnp_WT","Adnp_Input",cpms3sums_long$group)))
                                             
cpms3sums_long$ChIP <- ifelse(grepl("^H3K9me3",cpms3sums_long$group),"H3K9me3",
                                      ifelse(grepl("^Adnp2",cpms3sums_long$group),"ADNP2",
                                        ifelse(grepl("^Adnp",cpms3sums_long$group),"ADNP",     "other")))
cpms3sums_long$replicate <- ifelse(grepl("_r1",cpms3sums_long$sample),"r1","r2")
                                    

###ADNP2 Input and WT only
cpms2sumsAves_long_sel$group <- cpms2sumsAves_long_sel$sample
ggplot(cpms2sumsAves_long_sel[cpms2sumsAves_long_sel$group %in% c("Adnp2_Input","Adnp2_WT_ut"),],aes(x=group,y=cpm,fill=group)) + geom_bar(stat="identity") + facet_wrap(vars(repeat_name),scales="free",ncol=5) + theme_bw()  + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +  scale_fill_manual(values=c("yellow3","#7570b3")) +
  geom_errorbar(aes(ymin=cpm-cpm_sd, ymax=cpm+cpm_sd), width=.2,position=position_dodge(.9)) + 
  geom_jitter(data=cpms3sums_long[cpms3sums_long$sample %in% c("Input_WT_ut_r1","Input_WT_ut_r2", "Input_WT_ut_r3","Input_WT_ut_r4","Adnp2_WT_ut_r1","Adnp2_WT_ut_r2", "Adnp2_WT_ut_r3","Adnp2_WT_ut_r4"),],aes(x=group,y=cpm),col="darkgrey",position = position_jitter(0.1))
ggsave(file.path(topdir,plotdir,"Adnp2_WT_repeat_cpm_ChIP_Vs_Input_selected_barplots_4reps_sd.png"),height=9,width=10)
ggsave(file.path(topdir,plotdir,"Adnp2_WT_repeat_cpm_ChIP_Vs_Input_selected_barplots_4reps_sd.pdf"),height=9,width=10)

```





