---
title: "differentail binding analysis using EdeR"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
suppressPackageStartupMessages({
  library(tidyverse)
  library(MiniChip)
  library(GenomicFeatures)
  library(cowplot)
  library(edgeR)
  library(ggpointdensity)
  library(ggpubr)
  library(QuasR)
})

#parameters
topdir <-  "../"
datadir <- "data"
plotdir <- "plots"
genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38spike/mm10_hg38Spike_refSTAR/chrNameLength.txt"
repeat_file <-  "../mm10_GRCm38.p6_GencodeM23/mm10_GRCm38.p6.genome_repmask.bed"
gene_file <- "../release_M23/gencode.vM23.annotation.txdb.sqlite"
genome_fileH <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"

minMQS=0

options(ucscChromosomeNames=FALSE)
```

### load the repeats

```{r peaks}

#prepare repeat annotations from repeat masker run
repeats <- read.delim(repeat_file,header=FALSE)
names(repeats) <- c("chr","start","end","repeat_name","swScore","strand","percDiv","percDel","percIns","genome_left","repeat_class","rep_start","rep_end","rep_left","ID")
repeats <- repeats[complete.cases(repeats),]
reps <- makeGRangesFromDataFrame(repeats,
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("chr"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field=c("strand"),
                                 starts.in.df.are.0based=TRUE)

#remove chromosomes not in combined genome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
reps <- reps[overlapsAny(reps,chr_sizes,minoverlap = 1)==TRUE]
common_levels <- seqlevels(reps)[seqlevels(reps) %in% seqlevels(chr_sizes)]
reps <- keepSeqlevels(reps,value = common_levels)
#add a unique names column
names(reps) <- paste(reps$repeat_name,seqnames(reps),start(reps),end(reps),sep="_")

```


### load the bam files

```{r bamfiles}
#mouse bam files
bamNames <- c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2","Adnp2_PxVxL_ut_r1","Adnp2_PxVxL_ut_r2","Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2","Adnp2_PxVxL_Setdb1KD_r1", "Adnp2_PxVxL_Setdb1KD_r2",
              "H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2","H3K9me3_PxVxL_ut_r1","H3K9me3_PxVxL_ut_r2","H3K9me3_WT_Setdb1KD_r1","H3K9me3_WT_Setdb1KD_r2","H3K9me3_PxVxL_Setdb1KD_r1","H3K9me3_PxVxL_Setdb1KD_r2","Input_WT_ut_r1","Input_WT_ut_r2","InputK9_WT_ut_r1","InputK9_WT_ut_r2",
              "Adnp_WT_r1","Adnp_WT_r2","Adnp_PxVxL_r1","Adnp_PxVxL_r2","InputAdnp_WT_r1","InputAdnp_WT_r2")
bamFiles <- paste0(topdir,"/bam/",bamNames,"_multi.bam")

#human bam files
bamNamesH <- bamNames[1:20]
bamFilesH <- paste0(topdir,"/bam/spike/",bamNames[1:20],"_multi_spike.bam")
```

### count the number of paired-end reads in reps

```{r}
#write a table to read in samples for QUASR
write.table(data.frame(FileName=bamFiles[1:20],SampleName=bamNames[1:20]),file="QUASR.txt",sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#translate options
cl <- makeCluster(20)
selectReadPosition <- "start"
orientation <- "any"
readShiftSize <- "halfInsert"


#generate project
proj <- qAlign("QUASR.txt", genome=genome_file, paired = "fr", clObj = cl)

#generate counts matrices
countsPE <- qCount(proj, reps,
                 selectReadPosition= selectReadPosition, orientation = orientation, shift = readShiftSize, 
                 useRead="any", clObj = cl, mapqMin = minMQS)

#count the total mapped reads for human spike ins
write.table(data.frame(FileName=bamFilesH,SampleName=bamNames[1:20]),file="QUASR.txt",sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#generate project
proj <- qAlign("QUASR.txt", genome=genome_fileH, paired = "fr", clObj = cl)

mapped.readsH <- alignmentStats(proj)[,"mapped"]
names(mapped.readsH) <- bamNamesH
```

### count the number of single-end reads in reps

```{r}
#write a table to read in samples for QUASR
write.table(data.frame(FileName=bamFiles[21:26],SampleName=bamNames[21:26]),file="QUASR.txt",sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#translate options
cl <- makeCluster(20)
selectReadPosition <- "start"
orientation <- "any"
readShiftSize <- 80

#generate project
proj <- qAlign("QUASR.txt", genome=genome_file, paired = "no", clObj = cl)

#generate counts matrices
countsSE <- qCount(proj, reps,
                 selectReadPosition= selectReadPosition, orientation = orientation, shift = readShiftSize, 
                 useRead="any", clObj = cl, mapqMin = minMQS)

mapped.reads <- alignmentStats(proj)[,"mapped"]
names(mapped.reads) <- bamNames[21:26]
```

### combine counts and mapped reads

```{r}
counts <- data.frame(cbind(countsPE[,-1],countsSE[,-1]))
counts$ID <- row.names(counts)
mappedReads <- c(mapped.readsH,mapped.reads)
```

# add repeat info to counts

```{r}
#add repeat info to cpms
reps$ID <- names(reps)
reps$repclass <- ifelse(grepl("^DNA",reps$repeat_class),"DNA",
                              ifelse(grepl("RNA",reps$repeat_class),"RNA",
                                     ifelse(grepl("^LINE",reps$repeat_class),"LINE",
                                         ifelse(grepl("^LTR",reps$repeat_class),"LTR",
                                                ifelse(grepl("^SINE",reps$repeat_class),"SINE",
                                                       ifelse(grepl("Satellite",reps$repeat_class),"Satellite",
                                                              ifelse(reps$repeat_class=="Simple_repeat" | reps$repeat_class=="Low_complexity","Simple/low_complexity",
                                                                     ifelse(reps$repeat_class=="random","random","other"))))))))
counts2 <- left_join(counts,data.frame(mcols(reps)),by=c("ID"="ID"))
saveRDS(counts2,file.path(topdir,datadir,"per_repeat_per_sample_counts_Quasr_TMM.rds"))
```

#sum up counts per repeat_family, calculate cpms

```{r}
counts_sums <- counts2 %>% group_by(repclass,repeat_class,repeat_name) %>% summarize_if(is.numeric, sum, na.rm = TRUE)

counts_all <- as.matrix(counts_sums[,4:29])
row.names(counts_all) <- counts_sums$repeat_name

d <- DGEList(counts_all,lib.size=mappedReads)
d <- calcNormFactors(d,method="TMM")
cpms <- data.frame(cpm(d))
cpms$ID <- row.names(cpms)

#add repeat info to cpms
reps$ID <- names(reps)
reps$repclass <- ifelse(grepl("^DNA",reps$repeat_class),"DNA",
                              ifelse(grepl("RNA",reps$repeat_class),"RNA",
                                     ifelse(grepl("^LINE",reps$repeat_class),"LINE",
                                         ifelse(grepl("^LTR",reps$repeat_class),"LTR",
                                                ifelse(grepl("^SINE",reps$repeat_class),"SINE",
                                                       ifelse(grepl("Satellite",reps$repeat_class),"Satellite",
                                                              ifelse(reps$repeat_class=="Simple_repeat" | reps$repeat_class=="Low_complexity","Simple/low_complexity",
                                                                     ifelse(reps$repeat_class=="random","random","other"))))))))
rep.name2rep.class <- unique(data.frame(mcols(reps)[,c(1,13)]))
saveRDS(rep.name2rep.class,file.path(topdir,datadir,"repeat_name_2_repeat_class_table.rds"))

cpms2 <- left_join(cpms,rep.name2rep.class,by=c("ID"="repeat_name"))
saveRDS(cpms2,file.path(topdir,datadir,"per_repeat_family_per_sample_cpms_TMM_spike_in_norm.rds"))

```


### edgeR differential binding analysis for ADNP2 

```{r}
countsADNP2 <- as.matrix(counts_sums[,4:19])
row.names(countsADNP2) <- counts_sums$repeat_name

d <- DGEList(countsADNP2[,1:6],lib.size=mappedReads[1:6])
d <- calcNormFactors(d,method="TMM")

#generate group info
snames <- colnames(countsADNP2[,1:6])
group <- c("ADNP2_WT","ADNP2_WT","ADNP2_PxVxL","ADNP2_PxVxL",
           "ADNP2_WT_Setdb1DP","ADNP2_WT_Setdb1DP")

plotMDS(d, col = as.numeric(group))
mm <- model.matrix(~0 + group)

#dispersion
y <- estimateDisp(d,mm)
#model
fit <- glmQLFit(y, mm,prior.count=3)

contrasts <- list(makeContrasts(groupADNP2_PxVxL - groupADNP2_WT, levels = colnames(coef(fit))),
                  makeContrasts(groupADNP2_WT_Setdb1DP - groupADNP2_WT, levels = colnames(coef(fit)))
                  )

names(contrasts) <- c("ADNP2_PxVxL_vs_WT","ADNP2_WT_Setdb1DP_vs_WT")

# use the contrasts defined above to get p-values and fold-changes
res <- list()
for (i in seq_along(contrasts)){
  qlf <- glmQLFTest(fit, contrast=contrasts[[i]])
  res[[i]] <- data.frame(topTags(qlf,n=nrow(qlf$table),adjust.method = "BH"))
  res[[i]]$Contrast <- names(contrasts)[i]
  res[[i]]$ID <- row.names(res[[i]])
}
res2 <- do.call("rbind",res)
res2$logadjPval <- -log10(res2$FDR)
res2$regulated <- ifelse(res2$FDR < 0.05 & res2$logFC > log2(1),"up",
                         ifelse(res2$FDR < 0.05 & res2$logFC < -log2(1),"down","no"))
table(res2$regulated)
write.table(res2,file=file.path(topdir,datadir,"Adnp2_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_mq0.txt"),sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#Volcanoplot
vpa <- ggplot(res2,aes(x=logFC,y=logadjPval)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast))
vpa <- vpa + geom_vline(xintercept=c(-1,1),alpha=0.4) + geom_hline(yintercept = 2,alpha=0.4)
vpa
ggsave(file.path(topdir,plotdir,"Adnp2_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_volcanoPlot_mq0.pdf"),device="pdf",height=5,width=10)

#MAplot
ggplot(res2,aes(x=logCPM,y=logFC,col=regulated)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast)) + 
  theme_bw() + scale_size_manual(values=c(1,0.5,1)) + 
  scale_alpha_manual(values=c(1,0.5,1)) +scale_color_manual(values=c("#e41a1c","grey","#e41a1c"))
ggsave(file.path(topdir,plotdir,"Adnp2_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_MAPlot_mq0.pdf"),device="pdf",height=5,width=13)
```

### edgeR differential binding analysis for H3K9me3 

```{r}
countsADNP2 <- as.matrix(counts_sums[,4:19])
row.names(countsADNP2) <- counts_sums$repeat_name

d <- DGEList(countsADNP2[,7:12],lib.size=mappedReads[7:12])
d <- calcNormFactors(d,method="none")

#generate group info
snames <- colnames(countsADNP2[,7:12])
group <- c("H3K9me3_WT","H3K9me3_WT","H3K9me3_PxVxL","H3K9me3_PxVxL",
           "H3K9me3_WT_Setdb1DP","H3K9me3_WT_Setdb1DP")

plotMDS(d, col = as.numeric(group))
mm <- model.matrix(~0 + group)

#dispersion
y <- estimateDisp(d,mm)
#model
fit <- glmQLFit(y, mm,prior.count=3)

contrasts <- list(makeContrasts(groupH3K9me3_PxVxL - groupH3K9me3_WT, levels = colnames(coef(fit))),
                  makeContrasts(groupH3K9me3_WT_Setdb1DP - groupH3K9me3_WT, levels = colnames(coef(fit)))
                  )

names(contrasts) <- c("H3K9me3_PxVxL_vs_WT","H3K9me3_WT_Setdb1DP_vs_WT")

# use the contrasts defined above to get p-values and fold-changes
res <- list()
for (i in seq_along(contrasts)){
  qlf <- glmQLFTest(fit, contrast=contrasts[[i]])
  res[[i]] <- data.frame(topTags(qlf,n=nrow(qlf$table),adjust.method = "BH"))
  res[[i]]$Contrast <- names(contrasts)[i]
  res[[i]]$ID <- row.names(res[[i]])
}
res2 <- do.call("rbind",res)
res2$logadjPval <- -log10(res2$FDR)
res2$regulated <- ifelse(res2$FDR < 0.05 & res2$logFC > log2(1),"up",
                         ifelse(res2$FDR < 0.05 & res2$logFC < -log2(1),"down","no"))
table(res2$regulated)
write.table(res2,file=file.path(topdir,datadir,"H3K9me3_edgeR_comparison_on_repeat_families__WT_PxVxL_Setdb1DP_mq0.txt"),sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#Volcanoplot
vpa <- ggplot(res2,aes(x=logFC,y=logadjPval)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast))
vpa <- vpa + geom_vline(xintercept=c(-1,1),alpha=0.4) + geom_hline(yintercept = 2,alpha=0.4)
vpa
ggsave(file.path(topdir,plotdir,"H3K9me3_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_volcanoPlot_mq0.pdf"),device="pdf",height=5,width=10)

#MAplot
ggplot(res2,aes(x=logCPM,y=logFC,col=regulated)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast)) + 
  theme_bw() + scale_size_manual(values=c(1,0.5,1)) + 
  scale_alpha_manual(values=c(1,0.5,1)) +scale_color_manual(values=c("#e41a1c","grey","#e41a1c"))
ggsave(file.path(topdir,plotdir,"H3K9me3_edgeR_comparison_on_repeat_families__WT_PxVxL_Setdb1DP_MAPlot_mq0.pdf"),device="pdf",height=5,width=13)
```

### edgeR differential binding analysis for ADNP 

```{r}

countsADNP2 <- as.matrix(counts_sums[,4:19])
row.names(countsADNP2) <- counts_sums$repeat_name

d <- DGEList(countsADNP2[,13:16],lib.size=mappedReads[13:16])
d <- calcNormFactors(d,method="none")

#generate group info
snames <- colnames(countsADNP2[,13:16])
group <- c("ADNP_WT","ADNP_WT","ADNP_PxVxL","ADNP_PxVxL")

plotMDS(d, col = as.numeric(group))
mm <- model.matrix(~0 + group)

#dispersion
y <- estimateDisp(d,mm)
#model
fit <- glmQLFit(y, mm,prior.count=3)

contrasts <- list(makeContrasts(groupADNP_PxVxL - groupADNP_WT, levels = colnames(coef(fit)))
                  )

names(contrasts) <- c("ADNP_PxVxL_vs_WT")

# use the contrasts defined above to get p-values and fold-changes
res <- list()
for (i in seq_along(contrasts)){
  qlf <- glmQLFTest(fit, contrast=contrasts[[i]])
  res[[i]] <- data.frame(topTags(qlf,n=nrow(qlf$table),adjust.method = "BH"))
  res[[i]]$Contrast <- names(contrasts)[i]
  res[[i]]$ID <- row.names(res[[i]])
}
res2 <- do.call("rbind",res)
res2$logadjPval <- -log10(res2$FDR)
res2$regulated <- ifelse(res2$FDR < 0.1 & res2$logFC > log2(1),"up",
                         ifelse(res2$FDR < 0.1 & res2$logFC < -log2(1),"down","no"))
table(res2$regulated)
write.table(res2,file=file.path(topdir,datadir,"Adnp_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_mq0.txt"),sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#Volcanoplot
vpa <- ggplot(res2,aes(x=logFC,y=logadjPval)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast))
vpa <- vpa + geom_vline(xintercept=c(-1,1),alpha=0.4) + geom_hline(yintercept = 2,alpha=0.4)
vpa
ggsave(file.path(topdir,plotdir,"Adnp_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_volcanoPlot_mq0.pdf"),device="pdf",height=5,width=10)

#MAplot
ggplot(res2,aes(x=logCPM,y=logFC,col=regulated)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast)) + 
  theme_bw() + scale_size_manual(values=c(1,0.5,1)) + 
  scale_alpha_manual(values=c(1,0.5,1)) +scale_color_manual(values=c("#e41a1c","grey","#e41a1c"))
ggsave(file.path(topdir,plotdir,"Adnp_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_MAPlot_mq0.pdf"),device="pdf",height=5,width=13)
```

### correlation of ADNP2 changes with H3K9me3 levels (cpm) and differences (logFCs)

```{r}
all_cpms1 <- readRDS(file.path(topdir,datadir,"per_repeat_family_per_sample_cpms_TMM_spike_in_norm.rds"))

res_ADNP2 <- read.table(file.path(topdir,datadir,"Adnp2_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_mq0.txt"),header=TRUE)
res_K9 <- read.table(file.path(topdir,datadir,"H3K9me3_edgeR_comparison_on_repeat_families__WT_PxVxL_Setdb1DP_mq0.txt"),header=TRUE)

all_cpms <- data.frame(log2(all_cpms1[,1:16] + 3))
all_cpms$ID <- all_cpms1$ID

res <- left_join(res_ADNP2[res_ADNP2$Contrast=="ADNP2_PxVxL_vs_WT",],res_K9[res_K9$Contrast=="H3K9me3_PxVxL_vs_WT",],by="ID")
res <- left_join(res,all_cpms,by="ID")
res$H3K9me3_cpms <- apply(res[,c("H3K9me3_WT_ut_r1", "H3K9me3_WT_ut_r2")],1,mean)
  
p1 <- ggplot(res,aes(x=logFC.x,y=logFC.y)) + geom_pointdensity() + facet_wrap(vars(Contrast.x)) + stat_cor() + xlab("ADNP2 logFC") + ylab("H3K9me3 logFC") + theme_bw()
p1b <- ggplot(res,aes(x=logFC.x,y=H3K9me3_cpms)) + geom_pointdensity() + facet_wrap(vars(Contrast.x)) + stat_cor() + xlab("ADNP2 logFC") + ylab("H3K9me3 ave levels") + theme_bw()

res <- left_join(res_ADNP2[res_ADNP2$Contrast=="ADNP2_WT_Setdb1DP_vs_WT",],res_K9[res_K9$Contrast=="H3K9me3_WT_Setdb1DP_vs_WT",],by="ID")
res <- left_join(res,all_cpms,by="ID")
res$H3K9me3_cpms <- apply(res[,c("H3K9me3_WT_ut_r1", "H3K9me3_WT_ut_r2")],1,mean)
p2 <- ggplot(res,aes(x=logFC.x,y=logFC.y)) + geom_pointdensity() + facet_wrap(vars(Contrast.x)) + stat_cor() + xlab("ADNP2 logFC") + ylab("H3K9me3 logFC") + theme_bw()
p2b <- ggplot(res,aes(x=logFC.x,y=H3K9me3_cpms)) + geom_pointdensity() + facet_wrap(vars(Contrast.x)) + stat_cor() + xlab("ADNP2 logFC") + ylab("H3K9me3 ave levels") + theme_bw()

plot_grid(p1,p2, p1b, p2b)

ggsave(file.path(topdir,plotdir,"ADNP2_vs_H3K9me3_logFC_EdegR_comparison_per_repeat_family_WT_PxVxL_Setdb1DP_mq0.pdf"),device="pdf",height=8,width=10)
```


# which repeat families are regulaed in ADNP PxVxL?

```{r}
res_ADNP <- read.table(file.path(topdir,datadir,"Adnp_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_mq0.txt"),header=TRUE)
res_ADNP <-  left_join(res_ADNP,rep.name2rep.class,by=c("ID"="repeat_name"))
res_ADNP_sel <- res_ADNP[res_ADNP$repclass != "Simple/low_complexity" & res_ADNP$repclass != "RNA" & res_ADNP$repclass != "other",]
res_ADNP_sel <- res_ADNP_sel[complete.cases(res_ADNP_sel),]
na.omit(res_ADNP_sel[res_ADNP_sel$regulated=="up","ID"])
na.omit(res_ADNP_sel[res_ADNP_sel$regulated=="down","ID"])

res_ADNP_sel$regulated_class <- ifelse(res_ADNP_sel$regulated !="no",res_ADNP_sel$repclass,"not_regulated")
res_ADNP_sel$label <- ifelse(res_ADNP_sel$regulated !="no" & res_ADNP_sel$logFC > 1 & res_ADNP_sel$logCPM >6,res_ADNP_sel$ID,"")
res_ADNP_sel$label <- ifelse(res_ADNP_sel$regulated !="no" & res_ADNP_sel$logFC < -0.65 & res_ADNP_sel$logCPM >6,res_ADNP_sel$ID,res_ADNP_sel$label)

#MA plot colored by repclass
ggplot(res_ADNP_sel,aes(x=logCPM,y=logFC,col=regulated_class)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast)) + scale_color_manual(values = c("not_regulated" ="grey",
                                "DNA"="#E27069FF","RNA"="#DBA520FF","LINE"="#E93FBCFF","LTR"="#2E8B57FF",
                               "SINE"="#1175BBFF","Satellite"="#7852A9FF","other"= "#EF7215FF")) + geom_text_repel(aes(x = logCPM, 
                      y = logFC, 
                      label = label),size=3)
ggsave(file.path(topdir,plotdir,"Adnp_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_MAPlot_mq0_replcass_color_noRNAreps.pdf"),device="pdf",height=5,width=8)
```

# which repeat families are regulaed in ADNP2 PxVxL?

```{r}
res_ADNP2 <- read.table(file.path(topdir,datadir,"Adnp2_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_mq0.txt"),header=TRUE)
res_ADNP2 <-  left_join(res_ADNP2,rep.name2rep.class,by=c("ID"="repeat_name"))
res_ADNP2_sel <- res_ADNP2[res_ADNP2$repclass != "Simple/low_complexity"& res_ADNP2$repclass != "RNA" & res_ADNP2$repclass != "other",]
res_ADNP2_sel <- res_ADNP2_sel[complete.cases(res_ADNP2_sel),]
na.omit(res_ADNP2_sel[res_ADNP2_sel$regulated=="up","ID"])
na.omit(res_ADNP2_sel[res_ADNP2_sel$regulated=="down","ID"])

res_ADNP2_sel$regulated_class <- ifelse(res_ADNP2_sel$regulated !="no",res_ADNP2_sel$repclass,"not_regulated")
res_ADNP2_sel$label <- ifelse(res_ADNP2_sel$regulated !="no" & res_ADNP2_sel$logFC > 1 & res_ADNP2_sel$logCPM >10,res_ADNP2_sel$ID,"")
res_ADNP2_sel$label <- ifelse(res_ADNP2_sel$regulated !="no" & res_ADNP2_sel$logFC < -1 & res_ADNP2_sel$logCPM >10,res_ADNP2_sel$ID,res_ADNP2_sel$label)

#MA plot colored by repclass
ggplot(res_ADNP2_sel,aes(x=logCPM,y=logFC,col=regulated_class)) +geom_point() + theme_classic() +facet_wrap(vars(Contrast)) + scale_color_manual(values = c("not_regulated" ="grey",
                                "DNA"="#E27069FF","RNA"="#DBA520FF","LINE"="#E93FBCFF","LTR"="#2E8B57FF",
                               "SINE"="#1175BBFF","Satellite"="#7852A9FF","other"= "#EF7215FF")) + geom_text_repel(aes(x = logCPM, 
                      y = logFC, 
                      label = label),size=3)
ggsave(file.path(topdir,plotdir,"Adnp2_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_MAPlot_mq0_replcass_color_noRNAreps.pdf"),device="pdf",height=5,width=13)
```


#compare Adnp and Adnp2 logFCs 

```{r}
res <- left_join(res_ADNP_sel,res_ADNP2_sel[res_ADNP2_sel$Contrast=="ADNP2_PxVxL_vs_WT",],by="ID")
res$label_both <- ifelse(res$label.x=="",res$label.y,res$label.x)
ggplot(res,aes(x=logFC.y,y=logFC.x,col=repclass.x)) + geom_point() + theme_classic() + scale_color_manual(values = c(
                                "DNA"="#E27069FF","LINE"="#E93FBCFF","LTR"="#2E8B57FF",
                               "SINE"="#1175BBFF","Satellite"="#7852A9FF")) + geom_text_repel(aes(x = logFC.y, 
                      y = logFC.x, 
                      label = label_both),size=3) + stat_cor() + xlab(" ADNP2 PxVxL vs WT") + ylab(" ADNP PxVxL vs WT")
ggsave(file.path(topdir,plotdir,"Adnp2_versus_Adnp_edgeR_comparison_on_repeat_families_WT_PxVxL_Setdb1DP_MAPlot_mq0_replcass_color_noRNAreps.pdf"),device="pdf",height=6,width=8)
```

# barplots of selected repeats

#### summarize per repeat family: sum of cpms per repeat family, mean and sd across replicates, save file

```{r}
cpms2 <- readRDS(file.path(topdir,datadir,"per_repeat_family_per_sample_cpms_TMM_spike_in_norm.rds"))
cpms2sums <- rename(cpms2,ID="repeat_name")

cpms2sumsAves <- data.frame(repclass=cpms2sums$repclass,repeat_name=cpms2sums$repeat_name,
                            Adnp_WT=apply(cpms2sums[,c("Adnp_WT_r1","Adnp_WT_r2")],1,mean),
                            Adnp_PxVxL=apply(cpms2sums[,c("Adnp_PxVxL_r1","Adnp_PxVxL_r2")],1,mean),
                            
                            Adnp2_WT_ut=apply(cpms2sums[,c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2")],1,mean),
                            #Adnp2_WT_ut=apply(cpms2sums[,c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2","Adnp2_WT_ut_r3","Adnp2_WT_ut_r4")],1,mean),

                            Adnp2_PxVxL_ut=apply(cpms2sums[,c("Adnp2_PxVxL_ut_r1","Adnp2_PxVxL_ut_r2")],1,mean),
                            Adnp2_WT_Setdb1KD=apply(cpms2sums[,c("Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2")],1,mean),
                            Adnp2_PxVxL_Setdb1KD=apply(cpms2sums[,c("Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r2")],1,mean),
                            
                            H3K9me3_WT_ut=apply(cpms2sums[,c("H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2")],1,mean),
                            H3K9me3_WT_Setdb1KD=apply(cpms2sums[,c("H3K9me3_WT_Setdb1KD_r1","H3K9me3_WT_Setdb1KD_r2")],1,mean),
                            H3K9me3_PxVxL_Setdb1KD=apply(cpms2sums[,c("H3K9me3_PxVxL_Setdb1KD_r1","H3K9me3_PxVxL_Setdb1KD_r2")],1,mean),
                            H3K9me3_PxVxL_ut=apply(cpms2sums[,c("H3K9me3_PxVxL_ut_r1","H3K9me3_PxVxL_ut_r2")],1,mean),

                            Adnp_Input=apply(cpms2sums[,c("InputAdnp_WT_r1","InputAdnp_WT_r2")],1,mean),
                            Adnp2_Input=apply(cpms2sums[,c("Input_WT_ut_r1","Input_WT_ut_r2")],1,mean),
                            #Adnp2_Input=apply(cpms2sums[,c("Input_WT_ut_r1","Input_WT_ut_r2","Input_WT_ut_r3","Input_WT_ut_r4")],1,mean),

                            H3K9me3_Input=apply(cpms2sums[,c("InputK9_WT_ut_r1","InputK9_WT_ut_r2")],1,mean))

cpms2sumsSDs <- data.frame(repclass=cpms2sums$repclass,repeat_name=cpms2sums$repeat_name,
                             Adnp_WT=apply(cpms2sums[,c("Adnp_WT_r1","Adnp_WT_r2")],1,sd),
                            Adnp_PxVxL=apply(cpms2sums[,c("Adnp_PxVxL_r1","Adnp_PxVxL_r2")],1,sd),
                           
                            Adnp2_WT_ut=apply(cpms2sums[,c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2")],1,sd),
                            #Adnp2_WT_ut=apply(cpms2sums[,c("Adnp2_WT_ut_r1","Adnp2_WT_ut_r2","Adnp2_WT_ut_r3","Adnp2_WT_ut_r4")],1,sd),

                            Adnp2_PxVxL_ut=apply(cpms2sums[,c("Adnp2_PxVxL_ut_r1","Adnp2_PxVxL_ut_r2")],1,sd),
                            Adnp2_WT_Setdb1KD=apply(cpms2sums[,c("Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2")],1,sd),
                            Adnp2_PxVxL_Setdb1KD=apply(cpms2sums[,c("Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r1")],1,sd),
                            
                            H3K9me3_WT_ut=apply(cpms2sums[,c("H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2")],1,sd),
                            H3K9me3_WT_Setdb1KD=apply(cpms2sums[,c("H3K9me3_WT_Setdb1KD_r1","H3K9me3_WT_Setdb1KD_r2")],1,sd),
                            H3K9me3_PxVxL_Setdb1KD=apply(cpms2sums[,c("H3K9me3_PxVxL_Setdb1KD_r1","H3K9me3_PxVxL_Setdb1KD_r2")],1,sd),
                            H3K9me3_PxVxL_ut=apply(cpms2sums[,c("H3K9me3_PxVxL_ut_r1","H3K9me3_PxVxL_ut_r2")],1,sd),


                            Adnp_Input=apply(cpms2sums[,c("InputAdnp_WT_r1","InputAdnp_WT_r2")],1,sd),
                            Adnp2_Input=apply(cpms2sums[,c("Input_WT_ut_r1","Input_WT_ut_r2")],1,sd),
                            #Adnp2_Input=apply(cpms2sums[,c("Input_WT_ut_r1","Input_WT_ut_r2","Input_WT_ut_r3","Input_WT_ut_r4")],1,sd),

                            H3K9me3_Input=apply(cpms2sums[,c("InputK9_WT_ut_r1","InputK9_WT_ut_r2")],1,sd)
                            )

saveRDS(cpms2sumsAves,file.path(topdir,datadir,"per_repeat_family_per_group_mean_cpms_QuasR_TMM.rds"))
saveRDS(cpms2sumsSDs,file.path(topdir,datadir,"per_repeat_family_per_group_sd_cpms_QuasR_TMM.rds"))
```


### put mean and sd tables in long format and combine

```{r}
cpms2sumsAves <- readRDS(file.path(topdir,datadir,"per_repeat_family_per_group_mean_cpms_QuasR_TMM.rds"))
cpms2sumsSDs <- readRDS(file.path(topdir,datadir,"per_repeat_family_per_group_sd_cpms_QuasR_TMM.rds"))

cpms2sumsAves_long <- cpms2sumsAves %>% pivot_longer(cols=colnames(cpms2sumsAves)[3:ncol(cpms2sumsAves)],names_to = "sample",values_to="cpm")
cpms2sumsSDs_long <- cpms2sumsSDs %>% pivot_longer(cols=colnames(cpms2sumsSDs)[3:ncol(cpms2sumsSDs)],names_to = "sample",values_to="cpm_sd")
cpms2sumsAves_long <- cbind(cpms2sumsAves_long,cpms2sumsSDs_long[,4])

```

# barplots of individual enriched interesting repeats, including replicate points. split by ChIP type.

```{r}
cpms2sumsAves_long_sel <- cpms2sumsAves_long[cpms2sumsAves_long$repeat_name %in% c("IAPEz.int","MMERVK10C.int","B2_Mm1a","B2_Mm1t","B2_Mm2","B3","B3A","GSAT_MM","MurSatRep1","IMPB_01","RLTR10C","IAPLTR1a_Mm","IAPA_MM.int","L1MdA_I","Tigger9b","UCON11"),]
cpms2sumsAves_long_sel$ChIP <- ifelse(grepl("^H3K9me3",cpms2sumsAves_long_sel$sample),"H3K9me3",
                                      ifelse(grepl("^Adnp2",cpms2sumsAves_long_sel$sample),"ADNP2","ADNP"))
  
##########individual replicate data for dots##############

cpms3sums_long  <- cpms2sums[cpms2sums$repeat_name %in% c("IAPEz.int","MMERVK10C.int","B2_Mm1a","B2_Mm1t","B2_Mm2","B3","B3A","GSAT_MM","MurSatRep1","IMPB_01","RLTR10C","IAPLTR1a_Mm","IAPA_MM.int","L1MdA_I","Tigger9b","UCON11"),] %>% pivot_longer(cols=colnames(cpms2sums)[1:(ncol(cpms2sums)-2)],names_to = "sample",values_to="cpm")

#remove the extra (non-matched) replicates from ADNP2 data
cpms3sums_long <- cpms3sums_long[grepl("_r[0-2]",cpms3sums_long$sample),]
cpms3sums_long$group <- gsub("_r[0-9]","",cpms3sums_long$sample)
cpms3sums_long$group <- ifelse(cpms3sums_long$group=="InputK9_WT_ut","H3K9me3_Input",
                               ifelse(cpms3sums_long$group=="Input_WT_ut","Adnp2_Input",
                                      ifelse(cpms3sums_long$group=="InputAdnp_WT","Adnp_Input",cpms3sums_long$group)))
                                             
cpms3sums_long$ChIP <- ifelse(grepl("^H3K9me3",cpms3sums_long$group),"H3K9me3",
                                      ifelse(grepl("^Adnp2",cpms3sums_long$group),"ADNP2",
                                        ifelse(grepl("^Adnp",cpms3sums_long$group),"ADNP",     "other")))
cpms3sums_long$replicate <- ifelse(grepl("_r1",cpms3sums_long$sample),"r1","r2")
                                    

#H3K9me3
cpms2sumsAves_long_H3K9me3 <- cpms2sumsAves_long_sel[cpms2sumsAves_long_sel$ChIP=="H3K9me3",]
cpms2sumsAves_long_H3K9me3$sample <- factor(cpms2sumsAves_long_H3K9me3$sample,levels=c("H3K9me3_WT_ut","H3K9me3_PxVxL_ut", "H3K9me3_WT_Setdb1KD", "H3K9me3_PxVxL_Setdb1KD", "H3K9me3_Input"),labels=c("H3K9me3_WT_ut","H3K9me3_PxVxL_ut", "H3K9me3_WT_Setdb1KD", "H3K9me3_PxVxL_Setdb1KD", "H3K9me3_Input"))
  
p1 <- ggplot(cpms2sumsAves_long_H3K9me3,aes(x=sample,y=cpm)) + geom_bar(stat="identity",fill="#ebb400") + facet_wrap(vars(repeat_name),scales="free",ncol=5) + theme_classic()  + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  #geom_errorbar(aes(ymin=cpm-cpm_sd, ymax=cpm+cpm_sd), width=.2,position=position_dodge(.9)) + 
  geom_jitter(data=cpms3sums_long[cpms3sums_long$ChIP=="H3K9me3",],aes(x=group,y=cpm,shape=replicate),col="darkgrey",position = position_jitter(0.1))

#ADNP2
cpms2sumsAves_long_ADNP2 <- cpms2sumsAves_long_sel[cpms2sumsAves_long_sel$ChIP=="ADNP2",]
cpms2sumsAves_long_ADNP2$sample <- factor(cpms2sumsAves_long_ADNP2$sample,levels=c("Adnp2_WT_ut","Adnp2_PxVxL_ut", "Adnp2_WT_Setdb1KD", "Adnp2_PxVxL_Setdb1KD", "Adnp2_Input"),labels=c("Adnp2_WT_ut","Adnp2_PxVxL_ut", "Adnp2_WT_Setdb1KD", "Adnp2_PxVxL_Setdb1KD", "Adnp2_Input"))
  
p2 <- ggplot(cpms2sumsAves_long_ADNP2,aes(x=sample,y=cpm,fill=sample)) + geom_bar(stat="identity",fill="#6e377d") + facet_wrap(vars(repeat_name),scales="free",ncol=5) + theme_classic()  + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) + 
  #geom_errorbar(aes(ymin=cpm-cpm_sd, ymax=cpm+cpm_sd), width=.2,position=position_dodge(.9)) +
  geom_jitter(data=cpms3sums_long[cpms3sums_long$ChIP=="ADNP2",],aes(x=group,y=cpm,shape=replicate),col="darkgrey",position = position_jitter(0.1))

#ADNP
cpms2sumsAves_long_ADNP <- cpms2sumsAves_long_sel[cpms2sumsAves_long_sel$ChIP=="ADNP",]
cpms2sumsAves_long_ADNP$sample <- factor(cpms2sumsAves_long_ADNP$sample,levels=c("Adnp_WT","Adnp_PxVxL","Adnp_Input"),labels=c("Adnp_WT","Adnp_PxVxL","Adnp_Input"))
p3 <- ggplot(cpms2sumsAves_long_ADNP,aes(x=sample,y=cpm,fill=sample)) + geom_bar(stat="identity",fill="#008ca5") + facet_wrap(vars(repeat_name),scales="free",ncol=5) + theme_classic()  + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +
 # geom_errorbar(aes(ymin=cpm-cpm_sd, ymax=cpm+cpm_sd), width=.2,position=position_dodge(.9)) + 
  geom_jitter(data=cpms3sums_long[cpms3sums_long$ChIP=="ADNP",],aes(x=group,y=cpm,shape=replicate),col="darkgrey",position = position_jitter(0.1))
plot_grid(p1,p2,p3)
ggsave(file.path(topdir,plotdir,"Adnp_Adnp2_H3K9me3_PxVxL_Setdb1_repeat_cpm_ChIP_Vs_Input_selected_barplots_with_replicates_ChIP-split_Quasr_TMM.png"),height=20,width=20)
ggsave(file.path(topdir,plotdir,"Adnp_Adnp2_H3K9me3_PxVxL_Setdb1_repeat_cpm_ChIP_Vs_Input_selected_barplots_with_replicates_ChIP-split_QUasr_TMM.pdf"),height=20,width=20)

###ADNP2 Input and WT only
cpms2sumsAves_long_sel$group <- cpms2sumsAves_long_sel$sample
ggplot(cpms2sumsAves_long_sel[cpms2sumsAves_long_sel$group %in% c("Adnp2_Input","Adnp2_WT_ut"),],aes(x=group,y=cpm,fill=group)) + geom_bar(stat="identity") + facet_wrap(vars(repeat_name),scales="free",ncol=5) + theme_bw()  + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1)) +  scale_fill_manual(values=c("#b8b7b5","#6e377d")) +
  geom_errorbar(aes(ymin=cpm-cpm_sd, ymax=cpm+cpm_sd), width=.2,position=position_dodge(.9)) + 
  geom_jitter(data=cpms3sums_long[cpms3sums_long$sample %in% c("Input_WT_ut_r1","Input_WT_ut_r2", "Adnp2_WT_ut_r1","Adnp2_WT_ut_r2"),],aes(x=group,y=cpm),col="darkgrey",position = position_jitter(0.1))
ggsave(file.path(topdir,plotdir,"Adnp2_WT_repeat_cpm_ChIP_Vs_Input_selected_barplots_2reps_sd_QuasRTMM.png"),height=9,width=10)
ggsave(file.path(topdir,plotdir,"Adnp2_WT_repeat_cpm_ChIP_Vs_Input_selected_barplots_2reps_sd_QuasRTMM.pdf"),height=9,width=10)

```

