---
title: "Adnp and Adnp2 changing peak repeat element overlap"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
suppressPackageStartupMessages({
 library(tidyverse)
library(MiniChip)
library(ComplexHeatmap)
library(GenomicFeatures)
library(Biostrings)
library(BSgenome.Mmusculus.UCSC.mm10)
library(Rsamtools)
library(UpSetR)
})

#parameters
topdir <-  "../"


datadir <- "data"
plotdir <- "plots"
genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38spike/mm10_hg38Spike_refSTAR/chrNameLength.txt"
repeat_file <-  "../mm10_GRCm38.p6_GencodeM23/mm10_GRCm38.p6.genome_repmask.bed"
gene_file <- "../release_M23/gencode.vM23.annotation.txdb.sqlite"

options(ucscChromosomeNames=FALSE)

```

### load peaks and differential bing results

```{r}
ADNP2_peaks <- readRDS(file.path(topdir,datadir,"Adnp2_Peaks_filtered_1.2foldEnr.rds"))
ADNP_peaks <- readRDS(file.path(topdir,datadir,"Adnp_Peaks_filtered_1.2foldEnr.rds"))

res_ADNP2 <- read.table(file.path(topdir,datadir,"Adnp2_edgeR_comparison_WT_PxVxL_Setdb1DP.txt"),header=TRUE)
res_ADNP <- read.table(file.path(topdir,datadir,"Adnp_edgeR_comparison_WT_PxVxL_Setdb1DP.txt"),header=TRUE)

#add differential binding results to peak GRanges objects
ADNP2_peaks$PxVxL_regulated <- left_join(data.frame(ID=ADNP2_peaks$name),res_ADNP2[res_ADNP2$Contrast=="ADNP2_PxVxL_vs_WT",c("ID","regulated")],by="ID")[,2]
ADNP2_peaks$Setdb1_regulated <- left_join(data.frame(ID=ADNP2_peaks$name),res_ADNP2[res_ADNP2$Contrast=="ADNP2_WT_Setdb1DP_vs_WT",c("ID","regulated")],by="ID")[,2]
ADNP_peaks$PxVxL_regulated <- left_join(data.frame(ID=ADNP_peaks$name),res_ADNP[res_ADNP$Contrast=="ADNP_PxVxL_vs_WT",c("ID","regulated")],by="ID")[,2]
```

### prepare annotations

```{r}

#prepare gene annotations
txdb=loadDb(gene_file)
genes <- genes(txdb)
TSSs <- promoters(genes,upstream=300,downstream=0)
names(TSSs) <- matrix(unlist(strsplit(names(TSSs),".",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]
exons <- exons(txdb)
introns <- unlist(intronsByTranscript(txdb))


#prepare repeat annotations from repeat masker run
repeats <- read.delim(repeat_file,header=FALSE)
names(repeats) <- c("chr","start","end","repeat_name","swScore","strand","percDiv","percDel","percIns","genome_left","repeat_class","rep_start","rep_end","rep_left","ID")
repeats <- repeats[complete.cases(repeats),]
reps <- makeGRangesFromDataFrame(repeats,
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("chr"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field=c("strand"),
                                 starts.in.df.are.0based=TRUE)

#remove chromosomes not in combined genome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
reps <- reps[overlapsAny(reps,chr_sizes,minoverlap = 1)==TRUE]

rep.name2rep.class <- unique(data.frame(mcols(reps)[,c(1,7)]))
```

### general overlap with genes, TSSs, repeats

```{r}
reg <- c("up","down","no")
for (i in seq_along(reg)){
#ADNP2  
sel_ADNP2_peaks <- ADNP2_peaks[ADNP2_peaks$PxVxL_regulated==reg[i]]

ADNP2_4_upset <- data.frame(reps=as.integer(as.logical(overlapsAny(resize(sel_ADNP2_peaks,width=300,fix="center"),reps,ignore.strand=TRUE))),
exons=as.integer(as.logical(overlapsAny(resize(sel_ADNP2_peaks,width=300,fix="center"),exons,ignore.strand=TRUE))),
introns=as.integer(as.logical(overlapsAny(resize(sel_ADNP2_peaks,width=300,fix="center"),introns,ignore.strand=TRUE))),
TSSs=as.integer(as.logical(overlapsAny(resize(sel_ADNP2_peaks,width=300,fix="center"),TSSs,ignore.strand=TRUE))),
ADNP=as.integer(as.logical(overlapsAny(resize(sel_ADNP2_peaks,width=300,fix="center"),resize(ADNP_peaks,width=300,fix="center"),ignore.strand=TRUE))))

pdf(file.path(topdir,plotdir,sprintf("PxVxL_%s_regulated_Adnp2_peaks_overlaps_%d_reps_%d_peaks.pdf",reg[i],sum(ADNP2_4_upset$reps),nrow(ADNP2_4_upset))),height=5,width=5)
upset(ADNP2_4_upset, nsets = 5, nintersects = 10, set_size.show = TRUE, order.by= "freq")
dev.off()

#ADNP
sel_ADNP_peaks <- ADNP_peaks[ADNP_peaks$PxVxL_regulated==reg[i]]
ADNP_4_upset <- data.frame(reps=as.integer(as.logical(overlapsAny(resize(sel_ADNP_peaks,width=300,fix="center"),reps,ignore.strand=TRUE))),
exons=as.integer(as.logical(overlapsAny(resize(sel_ADNP_peaks,width=300,fix="center"),exons,ignore.strand=TRUE))),
introns=as.integer(as.logical(overlapsAny(resize(sel_ADNP_peaks,width=300,fix="center"),introns,ignore.strand=TRUE))),
TSSs=as.integer(as.logical(overlapsAny(resize(sel_ADNP_peaks,width=300,fix="center"),TSSs,ignore.strand=TRUE))),
ADNP2=as.integer(as.logical(overlapsAny(resize(sel_ADNP_peaks,width=300,fix="center"),resize(ADNP2_peaks,width=300,fix="center"),ignore.strand=TRUE))))

pdf(file.path(topdir,plotdir,sprintf("PxVxL_%s_regulated_Adnp_peaks_overlaps_%d_reps_%d_peaks.pdf",reg[i],sum(ADNP_4_upset$reps),nrow(ADNP_4_upset))),height=5,width=5)
upset(ADNP_4_upset, nsets = 5, nintersects = 10, set_size.show = TRUE, order.by= "freq")
dev.off()
}

```

### since many overlap TSSs, we add TSSs to the repeats table 

```{r}
mcols(reps) <- mcols(reps)[,c("repeat_name","repeat_class")]
mcols(TSSs) <- NULL
TSSs$repeat_name <- "TSS"
TSSs$repeat_class <- "TSS"
reps <- c(reps,TSSs)
rep.name2rep.class <- unique(data.frame(mcols(reps)[,c("repeat_name","repeat_class")]))

```

### select Adnp2 peaks and random peaks overlapping repeats

```{r}
reg <- c("up","down","no")
ADNP2_peaks_reglist <- list(0)
ADNP2_randompeaks_reglist <- list(0)

for (r in seq_along(reg)){
#select regulated peaks
  sel_ADNP2_peaks <- ADNP2_peaks[ADNP2_peaks$PxVxL_regulated==reg[r]]

  #seelct the repeats that overlap peaks
  reps.Adnp2 <- reps[overlapsAny(reps,resize(sel_ADNP2_peaks,width=300,fix="center"))]


  #generate a set of repeats that overlap n random peak sets
  npeaks <- 100
  reps.Adnp2.random <- list()
  for (i in 1:npeaks){
    random.peaks <- SimulatePeaks(length(sel_ADNP2_peaks),peak.widths = 300,chromosomeSizes = chrsizes_file)
    #select the repeats that overlap it
    reps.Adnp2.random[[i]] <- reps[overlapsAny(reps,resize(random.peaks,width=300,fix="center"))]
  }
  ADNP2_peaks_reglist[[r]] <- reps.Adnp2
  ADNP2_randompeaks_reglist[[r]] <- reps.Adnp2.random
}
names(ADNP2_peaks_reglist) <- reg
names(ADNP2_randompeaks_reglist) <- reg

```

### select Adnp peaks and random peaks overlapping repeats

```{r}
reg <- c("up","down","no")
ADNP_peaks_reglist <- list(0)
ADNP_randompeaks_reglist <- list(0)

for (r in seq_along(reg)){
#select regulated peaks
  sel_ADNP_peaks <- ADNP_peaks[ADNP_peaks$PxVxL_regulated==reg[r]]
#seelct the repeats that overlap peaks
reps.Adnp <- reps[overlapsAny(reps,resize(sel_ADNP_peaks,width=300,fix="center"))]

#generate a set of repeats that overlap n random peak sets
npeaks <- 100
reps.Adnp.random <- list()
for (i in 1:npeaks){
  random.peaks <- SimulatePeaks(length(sel_ADNP_peaks),peak.widths = 300,chromosomeSizes = chrsizes_file)
  #select the repeats that overlap it
  reps.Adnp.random[[i]] <- reps[overlapsAny(reps,resize(random.peaks,width=300,fix="center"))]
}
 ADNP_peaks_reglist[[r]] <- reps.Adnp
  ADNP_randompeaks_reglist[[r]] <- reps.Adnp.random
}
names(ADNP_peaks_reglist) <- reg
names(ADNP_randompeaks_reglist) <- reg
```

### Adnp2: comparison to repeat name annotation

```{r}
reg <- c("up","down","no")

for (r in seq_along(reg)){
  #extract the ADNP2 overlapping and random overlapping repeats from the list for the current regulated type
  reps.Adnp2 <- ADNP2_peaks_reglist[[r]]
  reps.Adnp2.random <- ADNP2_randompeaks_reglist[[r]]
  
#calculate the number of each repeat_name in each repeat set 
reps.tables <- list()
for (i in 1:npeaks){
  reps.tables[[i]] <- data.frame(table(reps.Adnp2.random[[i]]$repeat_name))
}

#merge the data.frames  
rep_table_random <- Reduce(
  function(x, y, ...) merge(x, y, all = TRUE, by="Var1",...),
  reps.tables
)
#replace all NAs with 0s
colnames(rep_table_random) <- c("repeat_name",paste("random",1:100,sep="."))
list(random.1=0,random.2=0)
replacelist <- vector(mode = "list", length = npeaks)
names(replacelist) <- colnames(rep_table_random)[-1]
replacelist <- lapply(replacelist,function(x){x <- 0})
rep_table_random2 <- replace_na(rep_table_random,replace = replacelist)

#calculate the mean and sd
rep_table_random3 <- data.frame(Var1=rep_table_random2$repeat_name,
                                mean.random=apply(rep_table_random2[,-1],1,mean),
                                sd.random=apply(rep_table_random2[,-1],1,sd)
)

#count them based on repeat name, compare in scatter plot
rep.name.distribution <- left_join(data.frame(table(reps.Adnp2$repeat_name)),rep_table_random3,
                                   by="Var1")
colnames(rep.name.distribution) <- c("repeat_name","Adnp2","random","sd")
rep.name.distribution$random <- replace_na(data = rep.name.distribution$random,replace = 0)
rep.name.distribution$sd <- replace_na(data = rep.name.distribution$sd,replace = 0)

#add rep class
rep.name.distribution <- left_join(rep.name.distribution,rep.name2rep.class,by="repeat_name")
#rep.name.distribution$repeat.class <- matrix(unlist(strsplit(rep.name.distribution$repeat_class,"/",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]

#save results as table
write.table(rep.name.distribution,file.path(topdir,datadir,sprintf("repeat_names_and_TSSs_at_Adnp2_%s_PxVxL_regulated_vs_random_peaks.txt",reg[r])),sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#plot
ggplot(rep.name.distribution,aes(y=Adnp2,x=random,col=repeat_class)) + geom_point() + theme_classic() +
  geom_text(aes(label=ifelse(Adnp2 > 75 & random < 50,as.character(repeat_name),'')),hjust=0.5,vjust=-0.7,size=2) +
  #xlim(0,1000) + ylim(0,1000) + 
  # scale_color_discrete()  + 
  geom_errorbarh(aes(xmax = random + sd, xmin = random - sd, height = 5)) +
  geom_abline(slope = 1, intercept = 0,col="grey",linetype="dotted")
ggsave(filename=file.path(topdir,plotdir,sprintf("repeat_names_and_TSSs_at_Adnp2_%s_PxVxL_regulated_vs_random_peaks.pdf",reg[r])),height=5, width=10,device = "pdf")
}
```

### Adnp: comparison to repeat name annotation

```{r}
reg <- c("up","down","no")

for (r in seq_along(reg)){
  #extract the ADNP2 overlapping and random overlapping repeats from the list for the current regulated type
  reps.Adnp <- ADNP_peaks_reglist[[r]]
  reps.Adnp.random <- ADNP_randompeaks_reglist[[r]]
  
#calculate the number of each repeat_name in each repeat set 
reps.tables <- list()
for (i in 1:npeaks){
  reps.tables[[i]] <- data.frame(table(reps.Adnp.random[[i]]$repeat_name))
}

#merge the data.frames  
rep_table_random <- Reduce(
  function(x, y, ...) merge(x, y, all = TRUE, by="Var1",...),
  reps.tables
)
#replace all NAs with 0s
colnames(rep_table_random) <- c("repeat_name",paste("random",1:100,sep="."))
list(random.1=0,random.2=0)
replacelist <- vector(mode = "list", length = npeaks)
names(replacelist) <- colnames(rep_table_random)[-1]
replacelist <- lapply(replacelist,function(x){x <- 0})
rep_table_random2 <- replace_na(rep_table_random,replace = replacelist)

#calculate the mean and sd
rep_table_random3 <- data.frame(Var1=rep_table_random2$repeat_name,
                                mean.random=apply(rep_table_random2[,-1],1,mean),
                                sd.random=apply(rep_table_random2[,-1],1,sd)
)

#count them based on repeat name, compare in scatter plot
rep.name.distribution <- left_join(data.frame(table(reps.Adnp$repeat_name)),rep_table_random3,
                                   by="Var1")
colnames(rep.name.distribution) <- c("repeat_name","Adnp","random","sd")
rep.name.distribution$random <- replace_na(data = rep.name.distribution$random,replace = 0)
rep.name.distribution$sd <- replace_na(data = rep.name.distribution$sd,replace = 0)

#add rep class
rep.name.distribution <- left_join(rep.name.distribution,rep.name2rep.class,by="repeat_name")
#rep.name.distribution$repeat.class <- matrix(unlist(strsplit(rep.name.distribution$repeat_class,"/",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]

#save results as table
write.table(rep.name.distribution,file.path(topdir,datadir,sprintf("repeat_names_and_TSSs_at_Adnp_%s_PxVxL_regulated_vs_random_peaks.txt",reg[r])),sep="\t",col.names=TRUE,row.names=FALSE,append=FALSE,quote=FALSE)

#plot
ggplot(rep.name.distribution,aes(y=Adnp,x=random,col=repeat_class)) + geom_point() + theme_classic() +
  geom_text(aes(label=ifelse(Adnp > 1000 & random < 10000,as.character(repeat_name),'')),hjust=0.5,vjust=-0.7,size=2) +
  #xlim(0,1000) + ylim(0,1000) + 
 # scale_color_discrete()  + 
  geom_errorbarh(aes(xmax = random + sd, xmin = random - sd, height = 5)) +
  geom_abline(slope = 1, intercept = 0,col="grey",linetype="dotted")
ggsave(filename=file.path(topdir,plotdir,sprintf("repeat_names_and_TSSs_at_Adnp_%s_PxVxL_regulated_vs_random_peaks.pdf",reg[r])),height=5, width=10,device = "pdf")

}
```




