---
title: "Adnp2 and Adnp motif search"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
suppressPackageStartupMessages({
 library(tidyverse)
library(MiniChip)
library(BSgenome.Mmusculus.UCSC.mm10)
  library(Biostrings)
})

#parameters
topdir <-  "../"
datadir <- "data"
plotdir <- "plots"
genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38spike/mm10_hg38Spike_refSTAR/chrNameLength.txt"
repeat_file <-  "../mm10_GRCm38.p6_GencodeM23/mm10_GRCm38.p6.genome_repmask.bed"
gene_file <- "../release_M23/gencode.vM23.annotation.txdb.sqlite"

options(ucscChromosomeNames=FALSE)

#make fasta file for MEME
mm10 <- readDNAStringSet(genome_file, format="fasta",
                 nrec=-1L, skip=0L, seek.first.rec=FALSE, use.names=TRUE)
names(mm10) <- matrix(unlist(strsplit(names(mm10)," ")),ncol=2,byrow=TRUE)[,1]


```

### load peaks 

```{r}
ADNP2_peaks <- readRDS(file.path(topdir,datadir,"Adnp2_Peaks_filtered_1.2foldEnr.rds"))

```

# sort peaks

```{r}
#sort peaks by score
ADNP2_peaks_sorted <-  ADNP2_peaks[order(ADNP2_peaks$score,decreasing=TRUE)]

```

# generate peak sequence files of selected peaks

```{r}
#ADNP2
peaknames="ADNP2_score_sorted"
width=500
topn=6315

#select the top ADNP2 peaks
peaks <- resize(ADNP2_peaks_sorted[1:topn], width = width, fix = "center")

#get their sequence
peaks.seq <- BSgenome::getSeq(mm10,peaks)
names(peaks.seq) <- names(peaks)
writeXStringSet(peaks.seq,file.path(topdir,fastadir,sprintf("%s.%sbp_top%s.fasta",peaknames,width,topn)),format = "fasta" )
```


# use MEME-ChIP to find motifs (all peaks)

```{bash}
module load MEME

meme-chip -oc meme/ADNP2_score_sorted.500bp_top6315_memechip -maxw 25 -meme-p 10 \
-db MEME/motif_databases/JASPAR/JASPAR2022_CORE_vertebrates_non-redundant_v2.meme \
fasta/ADNP2_score_sorted.500bp_top6315.fasta
```

