---
title: "Adnp2 repeat consensus mapping"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```


```{r libraries}
suppressPackageStartupMessages({
 library(tidyverse)
library(cowplot)
})

#parameters
topdir <-  "../"
datadir <- "data"
plotdir <- "plots"
genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38spike/mm10_hg38Spike_refSTAR/chrNameLength.txt"
repeat_file <-  "../mm10_GRCm38.p6_GencodeM23/mm10_GRCm38.p6.genome_repmask.bed"
gene_file <- "../release_M23/gencode.vM23.annotation.txdb.sqlite"

SEreadShift <- 80

options(ucscChromosomeNames=FALSE)
```

### prepare repeat masker annotations

```{r}

#prepare repeat annotations from repeat masker run
repeats <- read.delim(repeat_file,header=FALSE)
names(repeats) <- c("chr","start","end","repeat_name","swScore","strand","percDiv","percDel","percIns","genome_left","repeat_class","rep_start","rep_end","rep_left","ID")
repeats <- repeats[complete.cases(repeats),]
reps <- makeGRangesFromDataFrame(repeats,
                                 keep.extra.columns=TRUE,
                                 ignore.strand=FALSE,
                                 seqinfo=NULL,
                                 seqnames.field=c("chr"),
                                 start.field=c("start"),
                                 end.field=c("end"),
                                 strand.field=c("strand"),
                                 starts.in.df.are.0based=TRUE)

#remove chromosomes not in combined genome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
reps <- reps[overlapsAny(reps,chr_sizes,minoverlap = 1)==TRUE]
common_levels <- seqlevels(reps)[seqlevels(reps) %in% seqlevels(chr_sizes)]
reps <- keepSeqlevels(reps,value = common_levels)
#add a unique names column
names(reps) <- paste(reps$repeat_name,seqnames(reps),start(reps),end(reps),sep="_")
#save as bed file
reps.df <- data.frame(reps)
write.table(reps.df[,1:3],file=file.path(topdir,datadir,"all_repeats.bed"),sep="\t",col.names=FALSE,row.names=FALSE, append=FALSE, quote=FALSE)
```

# extract all reads that map to repeat sequences

```{bash}
module load SAMtools/1.10-foss-2019b
###module load BEDTools/2.27.1-foss-2018b
cd /tungstenfs/scratch/gbuehler/michi/Projects/Adnp2/manuscript/bam
repeat_bed="../data/all_repeats.bed"

for sample in Input_WT_ut_r1 Input_WT_ut_r2 Input_WT_ut_r3 Input_WT_ut_r4 Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4 Adnp2_PxVxL_Setdb1KD_r1 H3K9me3_WT_Setdb1KD_r1 Adnp2_PxVxL_Setdb1KD_r2 H3K9me3_WT_Setdb1KD_r2 Adnp2_PxVxL_ut_r1 H3K9me3_WT_ut_r1 InputK9_WT_ut_r1 Adnp2_PxVxL_ut_r2 H3K9me3_PxVxL_Setdb1KD_r1 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 InputK9_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 H3K9me3_PxVxL_ut_r1 Adnp2_WT_Setdb1KD_r2 H3K9me3_PxVxL_ut_r2
do
samtools view -b -@ 20 -L ${repeat_bed} ${sample}_multi.bam | samtools sort -@ 20 -n -o ${sample}_multi_reps_qsorted.bam
samtools fastq -@ 20 -1 ${sample}_multi_reps1.fastq -2 ${sample}_multi_reps2.fastq -s ${sample}_multi_reps.fastq ${sample}_multi_reps_qsorted.bam 
done
```

# map these reads to repeat consensus seqs

```{bash}
INDEX=../fasta/mouse_repeat_consensus_repmasker_220720
fastqdir=../bam

module purge
module load Bowtie2/2.3.5.1-GCC-8.3.0

#map
for sample in H3K9me3_PxVxL_Setdb1KD_r1 Input_WT_ut_r1 Input_WT_ut_r2 Input_WT_ut_r3 Input_WT_ut_r4 Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4 Adnp2_PxVxL_Setdb1KD_r1 H3K9me3_WT_Setdb1KD_r1 Adnp2_PxVxL_Setdb1KD_r2 H3K9me3_WT_Setdb1KD_r2 Adnp2_PxVxL_ut_r1 H3K9me3_WT_ut_r1 InputK9_WT_ut_r1 Adnp2_PxVxL_ut_r2 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 InputK9_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 H3K9me3_PxVxL_ut_r1 Adnp2_WT_Setdb1KD_r2 H3K9me3_PxVxL_ut_r2
do
bowtie2 -q -D 20 -R 3 -N 1 -L 20 -i S,1,0.50 -x $INDEX -1 $fastqdir/${sample}_multi_reps1.fastq -2 $fastqdir/${sample}_multi_reps2.fastq \
-S $fastqdir/${sample}.REPconsensus.sam --local -p 20 --no-unal 
done

module purge
module load SAMtools/1.10-foss-2019b

#convert to bam,sort and index
for sample in H3K9me3_PxVxL_Setdb1KD_r1 Input_WT_ut_r1 Input_WT_ut_r2 Input_WT_ut_r3 Input_WT_ut_r4 Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4 Adnp2_PxVxL_Setdb1KD_r1 H3K9me3_WT_Setdb1KD_r1 Adnp2_PxVxL_Setdb1KD_r2 H3K9me3_WT_Setdb1KD_r2 Adnp2_PxVxL_ut_r1 H3K9me3_WT_ut_r1 InputK9_WT_ut_r1 Adnp2_PxVxL_ut_r2 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 InputK9_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 H3K9me3_PxVxL_ut_r1 Adnp2_WT_Setdb1KD_r2 H3K9me3_PxVxL_ut_r2
do
samtools view -b -S $fastqdir/${sample}.REPconsensus.sam | samtools sort > $fastqdir/${sample}.REPconsensus.sorted.bam
samtools index $fastqdir/${sample}.REPconsensus.sorted.bam
done
```


# make bedgraphs (normalized to 1 Mio mouse reads)

```{bash}
module load BEDTools/2.27.1-foss-2018b
module load SAMtools/1.9-foss-2018b

for sample in H3K9me3_PxVxL_Setdb1KD_r1 Input_WT_ut_r1 Input_WT_ut_r2 Input_WT_ut_r3 Input_WT_ut_r4 Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4 Adnp2_PxVxL_Setdb1KD_r1 H3K9me3_WT_Setdb1KD_r1 Adnp2_PxVxL_Setdb1KD_r2 H3K9me3_WT_Setdb1KD_r2 Adnp2_PxVxL_ut_r1 H3K9me3_WT_ut_r1 InputK9_WT_ut_r1 Adnp2_PxVxL_ut_r2 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 InputK9_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 H3K9me3_PxVxL_ut_r1 Adnp2_WT_Setdb1KD_r2 H3K9me3_PxVxL_ut_r2
do
chromSize=../fasta/mouse_repeat_consensus_repmasker_220720.sizes

#normalisation factor for read counts 
n=$(samtools view -b bam/${sample}_multi.bam -F 4 -@ 10 | wc -l) 

scale=1000000
norm=$(echo ${n} ${scale} | awk '{print $2/($1*0.5)}')
          
# generate coverage files using pe information but only read pairs with inserts less than 1kb
       samtools view -h bam/${sample}.REPconsensus.sorted.bam -F 4 -@ 4 | awk 'substr($0,1,1)=="@" || ($9>= -1000 && $9<=1000)' | samtools view -b -@ 4 | bedtools genomecov -d -pc -ibam stdin -g ${chromSize} -scale ${norm} > bedgraph/${sample}_REPconsensus_multi.bedgraph 

done 
```


# combine coverage tracks for all samples and save

```{r}
samples <- c("Adnp2_WT_ut_r1", "Adnp2_WT_ut_r2", "Adnp2_WT_ut_r3", "Adnp2_WT_ut_r4", "Input_WT_ut_r1", "Input_WT_ut_r2", "Input_WT_ut_r3", "Input_WT_ut_r4","Adnp2_PxVxL_ut_r1","Adnp2_PxVxL_ut_r2","Adnp2_WT_Setdb1KD_r1","Adnp2_WT_Setdb1KD_r2","Adnp2_PxVxL_Setdb1KD_r1","Adnp2_PxVxL_Setdb1KD_r2","H3K9me3_WT_ut_r1","H3K9me3_WT_ut_r2","H3K9me3_PxVxL_Setdb1KD_r1","H3K9me3_WT_Setdb1KD_r1","H3K9me3_PxVxL_Setdb1KD_r2","H3K9me3_WT_Setdb1KD_r2","H3K9me3_PxVxL_ut_r1", "H3K9me3_PxVxL_ut_r2","InputK9_WT_ut_r1",    "InputK9_WT_ut_r2")
        

covTracksList <- list()
for (s in seq_along(samples)){
  covTracksList[[s]] <- read.table(file.path(topdir,"bedgraph",sprintf("%s_REPconsensus_multi.bedgraph",samples[s])),sep="\t",comment.char="^")
colnames(covTracksList[[s]]) <- c("repeat_name","position","cpm")
covTracksList[[s]]$sample <- samples[s]

}
covTracks <- do.call("rbind",covTracksList)
write.table(covTracks,file=file.path(topdir,datadir,"repeat_consensus_all_PE_coverage_tracks.txt"),sep="\t",col.names=TRUE,row.names=FALSE, append=FALSE, quote=FALSE)
```

# plot with ggplot 

```{r}
covTracks$ChIP <- ifelse(grepl("^Adnp2",covTracks$sample),"Adnp2",
                         ifelse(grepl("^H3K9me3",covTracks$sample),"H3K9me3", "Input"))
covTracks$genotype <- ifelse(grepl("WT_ut",covTracks$sample),"WT_ut",
                         ifelse(grepl("PxVxL_ut",covTracks$sample),"PxVxL_ut",
                                ifelse(grepl("WT_Setdb1KD",covTracks$sample),"WT_Setdb1KD","PxVxL_Setdb1KD")))      

covTracks$replicate <- ifelse(grepl("r1",covTracks$sample),"r1",
                         ifelse(grepl("r2",covTracks$sample),"r2",
                                ifelse(grepl("r3",covTracks$sample),"r3","r4")))                          

p1 <- ggplot(covTracks[covTracks$repeat_name %in% c("IAPEz-int#LTR/ERVK","IAPLTR1a_Mm#LTR/ERVK"),],aes(x=position,y=cpm,col=genotype,linetype=replicate)) + geom_smooth() + ylab("cpm") + xlab("position") + 
  theme_classic() + facet_grid(ChIP ~ repeat_name,scales="free") 

p2 <- ggplot(covTracks[covTracks$repeat_name%in% c("L1MdA_I_5end#LINE/L1","L1MdA_I_orf2#LINE/L1","L1MdA_I_3end#LINE/L1"),],aes(x=position,y=cpm,col=genotype,linetype=replicate)) + geom_smooth() + ylab("cpm") + xlab("position") + 
  theme_classic() + facet_grid(ChIP ~ repeat_name,scales="free") 

p3 <- ggplot(covTracks[covTracks$repeat_name%in% c("MMERVK10C#LTR/ERVK","RLTR10C#LTR/ERVK"),],aes(x=position,y=cpm,col=genotype,linetype=replicate)) + geom_smooth() + ylab("cpm") + xlab("position") + 
  theme_classic() + facet_grid(ChIP ~ repeat_name,scales="free")

plot_grid(p1,p2,p3)

ggsave(file.path(topdir,plotdir,"IAPEz_L1MdA_I_MMERVK10C_coverage_Adnp2_H3K9me3_Input_norm2mouse.pdf"),device="pdf",
       height=10,width=20)
```

