---
title: "generate bigwig tracks"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```
 
# generate bigwig files for paired-end data
 
```{bash}
module load BEDTools/2.27.1-foss-2018b
module load SAMtools/1.9-foss-2018b

for sample in Input_WT_ut_r1 Input_WT_ut_r2 Input_WT_ut_r3 Input_WT_ut_r4 Adnp2_WT_ut_r1 Adnp2_WT_ut_r2 Adnp2_WT_ut_r3 Adnp2_WT_ut_r4 Adnp2_PxVxL_Setdb1KD_r1 H3K9me3_WT_Setdb1KD_r1 Adnp2_PxVxL_Setdb1KD_r2 H3K9me3_WT_Setdb1KD_r2 Adnp2_PxVxL_ut_r1 H3K9me3_WT_ut_r1 InputK9_WT_ut_r1 Adnp2_PxVxL_ut_r2 H3K9me3_PxVxL_Setdb1KD_r2 H3K9me3_WT_ut_r2 InputK9_WT_ut_r2 Adnp2_WT_Setdb1KD_r1 H3K9me3_PxVxL_ut_r1 Adnp2_WT_Setdb1KD_r2 H3K9me3_PxVxL_ut_r2 
do
chromSize=../mm10_hg38spike/mm10_hg38Spike_refSTAR/chrNameLength.txt

#normalisation factor for read counts (based on human reads)
n=$(samtools view -b bam/${sample}_multi.bam -F 4 -@ 10 | wc -l) #mouse
scale=1000000
norm=$(echo ${n} ${scale} | awk '{print $2/($1*0.5)}')
          
# generate coverage files using pe information but only read pairs with inserts less than 1kb
samtools view -h bam/${sample}_multi.bam -F 4 -@ 20 | awk 'substr($0,1,1)=="@" || ($9>= -1000 && $9<=1000)' | samtools view -b -@ 20 | bedtools genomecov -bg -pc -ibam stdin -g ${chromSize} -scale ${norm} > bedgraph/${sample}_multi.bedgraph 
bedGraphToBigWig bedgraph/${sample}_multi.bedgraph ${chromSize} bigwig/${sample}_multi.bw 
done 
```
 
 