---
title: "assosication of ADNP2 peaks with regulated genes"
author: Michaela Schwaiger
output:
  html_document:
    toc: true
    toc_float: true
date: "`r Sys.Date()`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
options(width = 90L)
knitr::opts_chunk$set(echo = TRUE)
```



```{r}
suppressPackageStartupMessages({
  library(tidyverse)
  library(GenomicRanges)
  library(BSgenome.Mmusculus.UCSC.mm10)
  library(GenomicFeatures)
  library(cowplot)
})

#parameters
topdir <-  "../"
datadir <- "data"
plotdir <- "plots"

genome_file <-  "../mm10_GRCm38.p6_GencodeM23/GRCm38.p6.genome.fa"
chrsizes_file <-  "../mm10_hg38Spike_refSTAR/chrNameLength.txt"
repeat_file <-  "../mm10_GRCm38.p6_GencodeM23/mm10_GRCm38.p6.genome_repmask.bed"
gene_file <- "../release_M23/gencode.vM23.annotation.txdb.sqlite"
```

### load peaks

```{r}
ADNP2_peaks <- readRDS(file.path(topdir,datadir,"Adnp2_Peaks_filtered_1.2foldEnr.rds"))
ADNP_peaks <- readRDS(file.path(topdir,datadir,"Adnp_Peaks_filtered_1.2foldEnr.rds"))

peaks <- c(ADNP_peaks,ADNP2_peaks)

#remove peaks that extend over the edge of chromosome
chr_sizes <- read.table(chrsizes_file)
chr_sizes$start <- 1
chr_sizes <- makeGRangesFromDataFrame(chr_sizes,
                                   keep.extra.columns=TRUE,
                                   ignore.strand=TRUE,
                                   seqinfo=NULL,
                                   seqnames.field=c("V1"),
                                   start.field=c("start"),
                                   end.field=c("V2"),
                                   starts.in.df.are.0based=FALSE)
common_levels <- seqlevels(peaks)[seqlevels(peaks) %in% seqlevels(chr_sizes)]
peaks <- keepSeqlevels(peaks,value = common_levels)
```


### load regulated genes

```{r}
resWT <- read.table(file.path(topdir,datadir,"Adnp2WT_2i_14d_gene_TE_analysis_cloneblocked_genes_mq255.txt"),sep="\t",header=TRUE)
resWT$genotype <- "Adnp2_WT"
resKO <- read.table(file.path(topdir,datadir,"Adnp2KO_2i_14d_gene_TE_analysis_cloneblocked_genes_mq255.txt"),sep="\t",header=TRUE)
resKO$genotype <- "Adnp2_KO"
resKOWT <- read.table(file.path(topdir,datadir,"Adnp2KO_2i_14d_KO_vs-WT_gene_TE_analysis_genes_mq255.txt"),sep="\t",header=TRUE)
resKOWT$genotype <- "Adnp2_KO_vs_WT"

res2 <- rbind(resWT,resKO,resKOWT)


res_genes <- res2[grepl("^ENS",res2$GeneID),]
res_genes$GeneID <- matrix(unlist(strsplit(res_genes$GeneID,".",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]

contrasts <- unique(res_genes$Contrast)
```


# load gene annotation and extract TSS position

```{r}
#prepare gene annotations
txdb=loadDb(gene_file)
genes <- genes(txdb)
TSSs <- promoters(genes,upstream=1,downstream=0)
names(TSSs) <- matrix(unlist(strsplit(names(TSSs),".",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]
```

# find distance to closest TSS to each peak

```{r}
dist2TSS <- distanceToNearest(peaks,TSSs,ignore.strand=TRUE)

peaks2 <- peaks[from(dist2TSS)]
mcols(peaks2) <- cbind(mcols(peaks2),mcols(TSSs[to(dist2TSS)]))
peaks2$distance <- mcols(dist2TSS)$distance
```

# add the gene regulation info

```{r}
res2$Contrast_genotype <- paste(res2$genotype,res2$Contrast,sep="_")
res3 <- res2[,c("GeneID","Contrast_genotype",  "regulated")] %>% pivot_wider(id_cols = GeneID,names_from = Contrast_genotype,values_from = regulated)
res3$GeneID <- matrix(unlist(strsplit(res3$GeneID ,".",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]
peaks2$GeneID <- matrix(unlist(strsplit(peaks2$gene_id ,".",fixed=TRUE)),ncol=2,byrow=TRUE)[,1]

mcols(peaks2) <- left_join(data.frame(mcols(peaks2)),res3,by="GeneID")
saveRDS(peaks2,file=file.path(topdir,datadir,"ADNP2_peaks_2_regulated_genes.rds"))

p1 <- data.frame(mcols(peaks2[grep("Adnp2_",names(peaks2))])) %>% ggplot(aes(x=log2(distance+1),col=Adnp2_KO_vs_WT_Adnp2KO_vs_WT_dTAG_un)) + geom_density() + theme_classic() + ggtitle("Adnp2 peaks")
p2 <- data.frame(mcols(peaks2[grep("Adnp2_",names(peaks2))])) %>% ggplot(aes(x=log2(distance+1),col=Adnp2_KO_vs_WT_Adnp2KO_vs_WT_dTAG14d)) + geom_density() + theme_classic() + ggtitle("Adnp2 peaks")
p3 <- data.frame(mcols(peaks2)) %>% ggplot(aes(x=log2(distance+1),col=Adnp2_KO_vs_WT_Adnp2KO_dTAG14d_vs_WT_un)) + geom_density() + theme_classic() + ggtitle("all peaks")
p4 <- data.frame(mcols(peaks2[grep("Adnp2_",names(peaks2))])) %>% ggplot(aes(x=log2(distance+1),col=Adnp2_WT_dTAG14d_vs_un)) + geom_density() + theme_classic() + ggtitle("Adnp peaks")
p5 <- data.frame(mcols(peaks2[grep("Adnp2_",names(peaks2))])) %>% ggplot(aes(x=log2(distance+1),col=Adnp2_KO_dTAG14d_vs_un)) + geom_density() + theme_classic() + ggtitle("Adnp peaks")

plot_grid(p1,p2,p3,p4,p5)
ggsave(file.path(topdir,plotdir,"distance_of_peaks_to_regulated_genes.pdf"),height=8,width=20)

```
